// zotero-plugin-scaffold/dist/cli.js
import { l as logger } from "./shared/scaffold-replace-BTU0g6CT.mjs";
import { a as Release, i as ExitSignals, n as Test, o as Build, r as Serve, t as Config } from "./shared/scaffold-src-CQEdY9K7.mjs";
import process from "node:process";
import { readFile } from "node:fs/promises";
import { Command } from "commander";
import { pathExists } from "fs-extra";
import tinyUpdateNotifier from "tiny-update-notifier";
//#region package.json
var name = "zotero-plugin-scaffold";
var version = "0.8.6";
//#endregion
//#region src/utils/gitignore.ts
async function checkGitIgnore() {
	if (!pathExists(".git")) return;
	if (!pathExists(".gitignore")) {
		logger.warn("No .gitignore file found");
		return;
	}
	const contents = await readFile(".gitignore", "utf-8");
	const miss = [
		"node_modules",
		".env",
		".scaffold"
	].filter((ignore) => !contents.match(ignore));
	if (miss.length !== 0) logger.warn(`We recommend adding the following to your .gitignore file: ${miss.join(", ")}`);
}
//#endregion
//#region src/utils/updater.ts
function updateNotifier(name, version) {
	tinyUpdateNotifier({ pkg: {
		name,
		version
	} }).then((update) => {
		if (update) {
			const notify = () => {
				logger.newLine();
				logger.info(`New version of ${update.name} available!`);
				logger.info(`Update: ${update.current} → ${update.latest} (${update.type})`);
				logger.newLine();
			};
			ExitSignals.forEach((sig) => process.once(sig, notify));
		}
	}).catch();
}
//#endregion
//#region src/cli.ts
async function main() {
	updateNotifier(name, version);
	process.env.NODE_ENV ??= "development";
	const cli = new Command();
	cli.version(version).usage("<command> [options]");
	cli.command("build").description("Build the plugin").option("--dev", "Builds the plugin in dev mode").option("--dist <dir>", "The relative path for the new output directory (default: build)").action(async (options) => {
		process.env.NODE_ENV = options.dev ? "development" : "production";
		await runCommand(Build, { dist: options.dist });
	});
	cli.command("serve").alias("dev").description("Start development server").action(async (_options) => {
		await runCommand(Serve, {});
	});
	cli.command("test").description("Run tests").option("--abort-on-fail", "Abort the test suite on first failure").option("--exit-on-finish", "Exit the test suite after all tests have run").option("--no-watch", "Exit the test suite after all tests have run").action(async (options) => {
		process.env.NODE_ENV = "test";
		await runCommand(Test, { test: {
			abortOnFail: options.abortOnFail,
			watch: !options.exitOnFinish && options.watch
		} });
	});
	cli.command("create").description("Create the plugin template").action(async (_options) => {
		logger.error("The create not yet implemented");
	});
	cli.command("release").description("Release the plugin").argument("[version]", "Target version: major, minor, patch, pre*, or specify version").option("--preid <preid>", "ID for prerelease").option("-y, --yes", "Skip confirmation").action(async (version, options) => {
		process.env.NODE_ENV = "production";
		await runCommand(Release, { release: { bumpp: {
			release: version,
			preid: options.preid,
			confirm: !options.yes
		} } });
	});
	cli.arguments("<command>").action((cmd) => {
		cli.outputHelp();
		logger.error(`Unknown command name "${cmd}".`);
	});
	cli.parse();
}
async function runCommand(CommandClass, config) {
	const instance = new CommandClass(await Config.loadConfig(config));
	process.on("SIGINT", instance.exit.bind(instance));
	await instance.run();
}
async function mainWithErrorHandler() {
	main().then(() => {
		checkGitIgnore();
	}).catch(onError);
	process.on("uncaughtException", onError);
}
function onError(err) {
	logger.error(err);
	if (err.output) logger.log(err.output.stderr);
	process.exit(1);
}
//#endregion
export { mainWithErrorHandler as default, runCommand };
