// zotero-plugin-scaffold/dist/shared/scaffold-src-CQEdY9K7.mjs
import { a as parseRepoUrl, c as LOG_LEVEL, i as dateFormat, l as logger, n as replaceDefine, o as template, s as toArray } from "./scaffold-replace-BTU0g6CT.mjs";
import { loadConfig, setupDotenv } from "c12";
import { debounce, delay, escapeRegExp, kebabCase, mapValues, toMerged } from "es-toolkit";
import { copy, emptyDir, ensureDir, move, outputFile, outputJSON, pathExists, readJSON, readJsonSync, remove, writeJson } from "fs-extra/esm";
import { createHooks } from "hookable";
import process$1 from "node:process";
import { format, inspect, styleText } from "node:util";
import { isCI, isDebug, isLinux, isMacOS, isWindows } from "std-env";
import { glob, globSync } from "tinyglobby";
import path, { basename, dirname, extname, isAbsolute, join, resolve } from "node:path";
import { build, context } from "esbuild";
import { readFile, stat, writeFile } from "node:fs/promises";
import { Transformer, parse, serialize } from "@fluent/syntax";
import fs, { createReadStream, existsSync, lstatSync, promises, realpathSync, statSync } from "node:fs";
import { parseSync, printSync } from "@swc/core";
import { createHash } from "node:crypto";
import AdmZip from "adm-zip";
import { execSync, spawn } from "node:child_process";
import { ProgressEvent, versionBump } from "bumpp";
import "node:os";
import { URL as URL$1, fileURLToPath, pathToFileURL } from "node:url";
import assert from "node:assert";
import v8 from "node:v8";
import { Octokit } from "octokit";
import chokidar from "chokidar";
import net from "node:net";
import { Buffer } from "node:buffer";
import domain from "node:domain";
import EventEmitter from "node:events";
import { Xvfb } from "xvfb-ts";
import http from "node:http";
//#region src/config.ts
/**
* Helper for user define configuration.
*/
function defineConfig(userConfig) {
	return userConfig;
}
/**
* Loads config
* @param [overrides] Highest Priority Configuration.
* @returns Config with userDefined and defaultConfig merged.
*/
async function loadConfig$1(overrides) {
	return resolveConfig$1((await loadConfig({
		name: "zotero-plugin",
		dotenv: true,
		packageJson: true,
		defaults: getDefaultConfig$1(),
		overrides
	})).config);
}
function resolveConfig$1(config) {
	logger.setLogLevel(config.logLevel);
	const pkgUser = readJsonSync("package.json", { encoding: "utf-8" });
	const { name, version } = pkgUser;
	const { owner, repo } = parseRepoUrl(pkgUser.repository?.url);
	config.name ||= name;
	config.id ||= config.name;
	config.namespace ||= config.name;
	config.xpiName ||= kebabCase(config.name);
	config.build.prefs.prefix ||= pkgUser.config.prefsPrefix || `extensions.${config.namespace}`;
	const isPreRelease = version.includes("-");
	const templateData = {
		owner,
		repo,
		version,
		isPreRelease,
		updateJson: isPreRelease ? "update-beta.json" : "update.json",
		xpiName: config.xpiName,
		buildTime: dateFormat("YYYY-mm-dd HH:MM:SS", /* @__PURE__ */ new Date())
	};
	config.updateURL = template(config.updateURL, templateData);
	config.xpiDownloadLink = template(config.xpiDownloadLink, templateData);
	config.build.define = mapValues(config.build.define, (v) => template(v, templateData));
	config.release.github.repository = template(config.release.github.repository, templateData);
	const hooks = createHooks();
	hooks.addHooks(config.build.hooks);
	hooks.addHooks(config.server.hooks);
	hooks.addHooks(config.release.hooks);
	hooks.addHooks(config.test.hooks);
	return {
		...config,
		pkgUser,
		version,
		hooks,
		logger,
		templateData
	};
}
const defaultConfig = {
	source: "src",
	watchIgnore: [],
	dist: ".scaffold/build",
	name: "",
	id: "",
	namespace: "",
	xpiName: "",
	xpiDownloadLink: "https://github.com/{{owner}}/{{repo}}/releases/download/v{{version}}/{{xpiName}}.xpi",
	updateURL: "https://github.com/{{owner}}/{{repo}}/releases/download/release/{{updateJson}}",
	build: {
		assets: "addon/**/*.*",
		define: {},
		fluent: {
			prefixFluentMessages: true,
			prefixLocaleFiles: true,
			ignore: [],
			dts: "typings/i10n.d.ts"
		},
		prefs: {
			prefix: "",
			prefixPrefKeys: true,
			dts: "typings/prefs.d.ts"
		},
		esbuildOptions: [],
		makeManifest: {
			enable: true,
			template: {
				manifest_version: 2,
				name: "",
				version: "",
				applications: { zotero: {
					id: "",
					update_url: ""
				} }
			}
		},
		makeUpdateJson: {
			updates: [],
			hash: true
		},
		hooks: {}
	},
	server: {
		devtools: true,
		startArgs: [],
		prefs: {},
		asProxy: false,
		prebuild: true,
		createProfileIfMissing: true,
		hooks: {}
	},
	addonLint: {},
	release: {
		bumpp: {
			release: "prompt",
			preid: "beta",
			confirm: true,
			execute: "",
			all: false,
			commit: "chore(publish): release v%s",
			noVerify: false,
			tag: "v%s"
		},
		changelog: "",
		github: {
			enable: "ci",
			repository: "{{owner}}/{{repo}}",
			updater: "release",
			comment: false,
			releaseNote: (ctx) => {
				return ctx.release.changelog;
			}
		},
		hooks: {}
	},
	test: {
		entries: "test",
		prefs: {},
		mocha: { timeout: 1e4 },
		abortOnFail: false,
		headless: false,
		startupDelay: 1e3,
		waitForPlugin: "() => true",
		watch: true,
		hooks: {}
	},
	logLevel: "INFO"
};
const getDefaultConfig$1 = () => defaultConfig;
//#endregion
//#region src/core/base.ts
var Base = class {
	ctx;
	constructor(ctx) {
		this.ctx = ctx;
	}
	get logger() {
		return this.ctx.logger;
	}
};
//#endregion
//#region src/core/builder/assets.ts
const DEFAULT_IGNORE = ["node_modules", ".git"];
async function copyAssets(source, dist, assets) {
	const sourceArr = toArray(source);
	const paths = await glob(assets, { ignore: [...DEFAULT_IGNORE, dist] });
	for (const file of paths) {
		const newPath = getNewPath(sourceArr, dist, file);
		await copy(file, newPath);
		logger.debug(`Copy ${file} to ${newPath}`);
	}
}
function getNewPath(sources, dist, path) {
	const pattern = new RegExp(sources.join("|"));
	return `${dist}/addon/${path.replace(pattern, "")}`;
}
//#endregion
//#region src/core/builder/esbuild.ts
function resolveConfig(dist, esbuildOptions) {
	const distAbsolute = resolve(dist);
	return esbuildOptions.map((option, i) => {
		if (option.outfile && !resolve(option.outfile).startsWith(distAbsolute)) {
			logger.debug(`'outfile' of esbuildOptions[${i}] is not in dist folder, it will be overwritten.`);
			option.outfile = `${dist}/${option.outfile}`;
		}
		if (option.outdir && !resolve(option.outdir).startsWith(distAbsolute)) {
			logger.debug(`'outdir' of esbuildOptions[${i}] is not in dist folder, it will be overwritten.`);
			option.outdir = `${dist}/${option.outdir}`;
		}
		return option;
	});
}
async function esbuild(dist, esbuildOptions) {
	if (esbuildOptions.length === 0) return;
	const options = resolveConfig(dist, esbuildOptions);
	return await Promise.all(options.map((esbuildOption) => build(esbuildOption)));
}
//#endregion
//#region src/core/builder/fluent.ts
async function buildLocale(dist, namespace, options) {
	const ignores = toArray(options.ignore);
	const localeNames = await getLocales(dist);
	const messageManager = new MessageManager(ignores);
	await Promise.all(localeNames.map(async (locale) => {
		const paths = await glob(`${dist}/addon/locale/${locale}/**/*.ftl`);
		await Promise.all(paths.map(async (path) => {
			const fm = new FluentManager();
			await fm.read(path);
			messageManager.addMessages(locale, fm.getMessages());
			if (options.prefixFluentMessages) {
				fm.prefixMessages(namespace);
				await fm.write(path);
			}
			if (options.prefixLocaleFiles) {
				const newPath = `${dirname(path)}/${namespace}-${basename(path)}`;
				await move(path, newPath);
				logger.debug(`Renamed FTL: ${path} → ${newPath}`);
			}
		}));
	}));
	const htmlPaths = await glob([`${dist}/addon/**/*.xhtml`, `${dist}/addon/**/*.html`]);
	await Promise.all(htmlPaths.map(async (htmlPath) => {
		const { processedContent, foundMessages } = processHTMLFile(await readFile(htmlPath, "utf-8"), namespace, messageManager.getFTLMessages(), ignores, htmlPath);
		messageManager.addMessages("html", foundMessages);
		if (options.prefixFluentMessages) await writeFile(htmlPath, processedContent);
	}));
	messageManager.validateMessages();
	if (options.dts) {
		const dtsContent = generateFluentDts([...messageManager.getFTLMessages()]);
		await outputFile(options.dts, dtsContent, "utf-8");
	}
}
async function getLocales(dist) {
	return (await glob(`${dist}/addon/locale/*`, { onlyDirectories: true })).map((p) => basename(p));
}
var FluentManager = class {
	source;
	resource;
	messages = [];
	constructor() {}
	parse(source) {
		this.source = source;
		this.resource = parse(source, {});
	}
	async read(path) {
		const content = await readFile(path, "utf-8");
		this.parse(content);
	}
	getMessages() {
		if (!this.resource) throw new Error("Resource must be parsed first.");
		this.messages.length = 0;
		this.messages.push(...this.resource.body.filter((entry) => entry.type === "Message").map((message) => message.id.name));
		return this.messages;
	}
	prefixMessages(namespace) {
		if (!this.resource) throw new Error("Resource must be parsed before applying prefix.");
		new FluentTransformer(namespace).genericVisit(this.resource);
	}
	serialize() {
		if (!this.resource) throw new Error("Resource not parsed. Cannot serialize.");
		return serialize(this.resource, {});
	}
	async write(path) {
		if (this.serialize() !== this.source) await writeFile(path, this.serialize());
	}
};
var FluentTransformer = class extends Transformer {
	constructor(prefix) {
		super();
		this.prefix = prefix;
	}
	needsPrefix(name) {
		return !!this.prefix && !name.startsWith(this.prefix);
	}
	visitMessage(node) {
		if (this.needsPrefix(node.id.name)) node.id.name = `${this.prefix}-${node.id.name}`;
		return this.genericVisit(node);
	}
	visitMessageReference(node) {
		if (this.needsPrefix(node.id.name)) node.id.name = `${this.prefix}-${node.id.name}`;
		return this.genericVisit(node);
	}
};
var MessageManager = class {
	ftlMessages = /* @__PURE__ */ new Map();
	htmlMessages = /* @__PURE__ */ new Set();
	ignores;
	constructor(ignores) {
		this.ignores = ignores;
	}
	addMessages(target, messages) {
		if (target === "html") messages.forEach((msg) => this.htmlMessages.add(msg));
		else {
			let ftlLocaleMessages = this.ftlMessages.get(target);
			if (!ftlLocaleMessages) {
				ftlLocaleMessages = /* @__PURE__ */ new Set();
				this.ftlMessages.set(target, ftlLocaleMessages);
			}
			messages.forEach((msg) => ftlLocaleMessages.add(msg));
		}
	}
	validateMessages() {
		this.htmlMessages.forEach((msg) => {
			if (this.ignores.includes(msg)) return;
			const missingLocales = [...this.ftlMessages.entries()].filter(([_, messages]) => !messages.has(msg)).map(([locale]) => locale);
			if (missingLocales.length > 0) logger.warn(`I10N id ${styleText("blue", msg)} missing in locale: ${missingLocales.join(", ")}`);
		});
	}
	getFTLMessages() {
		const allMessages = /* @__PURE__ */ new Set();
		this.ftlMessages.forEach((messages) => messages.forEach((msg) => allMessages.add(msg)));
		return allMessages;
	}
	getFTLMessagesByLocale(locale) {
		return this.ftlMessages.get(locale) || /* @__PURE__ */ new Set();
	}
	getHTMLMessages() {
		return this.htmlMessages;
	}
};
function processHTMLFile(content, namespace, allMessages, ignores, filePath) {
	const foundMessages = /* @__PURE__ */ new Set();
	const L10N_PATTERN = new RegExp(`(data-l10n-id)="((?!${namespace})\\S*)"`, "g");
	return {
		processedContent: content.replace(L10N_PATTERN, (match, attr, id) => {
			foundMessages.add(id);
			if (ignores.includes(id)) {
				logger.debug(`Skipped ignored ID: ${styleText("blue", id)} in ${styleText("gray", filePath)}`);
				return match;
			}
			if (!allMessages.has(id)) {
				logger.warn(`I10N id ${styleText("blue", id)} in path ${styleText("gray", filePath)} does not exist in any locale, skip renaming it.`);
				return match;
			}
			return `${attr}="${namespace}-${id}"`;
		}),
		foundMessages: [...foundMessages]
	};
}
function generateFluentDts(messages) {
	return `// Generated by zotero-plugin-scaffold
/* prettier-ignore */
/* eslint-disable */
// @ts-nocheck
export type FluentMessageId =
${messages.toSorted().map((id) => `  | '${id}'`).join("\n")};
`;
}
//#endregion
//#region src/core/builder/manifest.ts
async function buildManifest(ctx) {
	if (!ctx.build.makeManifest.enable) return;
	const { name, id, updateURL, dist, version } = ctx;
	const manifestPath = `${dist}/addon/manifest.json`;
	const userData = await pathExists(manifestPath) ? await readJSON(manifestPath) : {};
	const data = toMerged(userData, {
		...userData,
		name: userData.name || name,
		version: userData.version || version,
		manifest_version: 2,
		applications: { zotero: {
			id,
			update_url: updateURL
		} }
	});
	logger.debug(`manifest: ${JSON.stringify(data, null, 2)}`);
	outputJSON(manifestPath, data, { spaces: 2 });
}
//#endregion
//#region src/utils/number.ts
/**
* Check if a number is a C++ integer.
*
* -2147483648 <= n <= 2147483647
*/
function is32BitNumber(n) {
	return Number.isInteger(n) && n >= -2147483648 && n <= 2147483647;
}
//#endregion
//#region src/utils/prefs-manager.ts
var PrefsManager = class {
	namespace;
	prefs = {};
	constructor(namespace) {
		this.namespace = namespace;
	}
	/**
	* Parse Method 3 - Using AST
	*/
	parse(content) {
		const _map = {};
		const ast = parseSync(content, { syntax: "ecmascript" });
		for (const node of ast.body) {
			if (node.type !== "ExpressionStatement" || node.expression.type !== "CallExpression" || node.expression.callee.type !== "Identifier" || node.expression.callee.value !== this.namespace || node.expression.arguments.length !== 2) throw new Error("Invalid prefs.js file.");
			const [arg1, arg2] = node.expression.arguments;
			if (arg1.expression.type !== "StringLiteral") throw new Error("Invalid prefs.js file - unsupported key type.");
			const key = arg1.expression.value.trim();
			let value;
			switch (arg2.expression.type) {
				case "StringLiteral":
				case "NumericLiteral":
				case "BooleanLiteral":
					value = arg2.expression.value;
					break;
				case "UnaryExpression":
					if (arg2.expression.argument.type !== "NumericLiteral") throw new Error("Invalid prefs.js file - unsupported value type.");
					if (arg2.expression.operator === "-") value = -arg2.expression.argument.value;
					else if (arg2.expression.operator === "+") value = arg2.expression.argument.value;
					else throw new Error("Invalid prefs.js file - unsupported value type.");
					break;
				case "TemplateLiteral":
					value = arg2.expression.quasis[0]?.cooked ?? "";
					break;
				default: throw new Error("Invalid prefs.js file - unsupported value type.");
			}
			_map[key] = value;
		}
		return _map;
	}
	/**
	* Parse Method 1 - Using RegExp
	* @deprecated
	*/
	parseByRegExp(content) {
		const _map = {};
		const matches = content.matchAll(/^(pref|user_pref)\s*\(\s*["']([^"']+)["']\s*,\s*(.+)\s*,?\s*\)\s*;?$/gm);
		for (const match of matches) {
			const key = match[2].trim();
			const value = match[3].trim();
			_map[key] = this.cleanValue(value);
		}
		return _map;
	}
	/**
	* Parse Method 2 - Using eval
	* @deprecated
	*/
	cleanValue(value) {
		if (value === "true") return true;
		else if (value === "false") return false;
		else if (!Number.isNaN(Number(value))) return Number(value);
		else if (/^["'](.*)["']$/.test(value)) return value.replace(/^["'](.*)["']$/, "$1");
		else return value;
	}
	/**
	* Render Method 2 - Using swc
	*/
	render() {
		const span = {
			start: 0,
			end: 0,
			ctxt: 0
		};
		function getExpression(value) {
			switch (typeof value) {
				case "string": return {
					type: "StringLiteral",
					span,
					value
				};
				case "boolean": return {
					type: "BooleanLiteral",
					span,
					value
				};
				case "number":
					if (value < 0) return {
						type: "UnaryExpression",
						span,
						operator: "-",
						argument: {
							type: "NumericLiteral",
							span,
							value: Math.abs(value)
						}
					};
					return {
						type: "NumericLiteral",
						span,
						value
					};
				default: throw new Error(`Unsupported value type: ${typeof value}`);
			}
		}
		const { code } = printSync({
			type: "Module",
			span,
			body: Object.entries(this.prefs).map(([key, value]) => ({
				type: "ExpressionStatement",
				span,
				expression: {
					type: "CallExpression",
					span,
					ctxt: 0,
					callee: {
						type: "Identifier",
						span,
						ctxt: 0,
						value: this.namespace,
						optional: false
					},
					arguments: [{ expression: getExpression(key) }, { expression: getExpression(value) }]
				}
			}))
		});
		return code;
	}
	/**
	* Render Method 1 - Using string
	* @deprecated
	*/
	renderByString() {
		return Object.entries(this.prefs).map(([key, value]) => {
			const _v = typeof value === "string" ? `"${value.replaceAll("\\", "\\\\").replaceAll("\"", "\\\"")}"` : value;
			return `${this.namespace}("${key}", ${_v});`;
		}).join("\n");
	}
	async read(path) {
		const content = await readFile(path, "utf-8");
		const map = this.parse(content);
		this.setPrefs(map);
	}
	async write(path) {
		await outputFile(path, this.render(), "utf-8");
		logger.debug("The prefs.js has been modified.");
	}
	setPref(key, value) {
		if (value === null || value === void 0) {
			if (key in this.prefs) delete this.prefs[key];
			return;
		}
		this.prefs[key] = value;
	}
	setPrefs(prefs) {
		Object.entries(prefs).forEach(([key, value]) => {
			this.setPref(key, value);
		});
	}
	getPref(key) {
		return this.prefs[key] ?? void 0;
	}
	getPrefs() {
		return this.prefs;
	}
	clearPrefs() {
		this.prefs = {};
	}
	getPrefsWithPrefix(prefix) {
		const _prefs = {};
		for (const pref in this.prefs) if (pref.startsWith(prefix)) _prefs[pref] = this.prefs[pref];
		else _prefs[`${prefix}.${pref}`] = this.prefs[pref];
		return _prefs;
	}
	getPrefsWithoutPrefix(prefix) {
		const _prefs = {};
		for (const pref in this.prefs) _prefs[pref.replace(`${prefix}.`, "")] = this.prefs[pref];
		return _prefs;
	}
};
//#endregion
//#region src/core/builder/prefs.ts
async function buildPrefs(dist, options) {
	const { dts, prefixPrefKeys, prefix } = options;
	if (!prefixPrefKeys && !dts) return;
	const prefsFilePath = join(dist, "addon", "prefs.js");
	if (!existsSync(prefsFilePath)) return;
	const prefsManager = new PrefsManager("pref");
	await prefsManager.read(prefsFilePath);
	const prefsWithPrefix = prefsManager.getPrefsWithPrefix(prefix);
	const prefsWithoutPrefix = prefsManager.getPrefsWithoutPrefix(prefix);
	const prefs = prefsManager.getPrefs();
	Object.entries(prefs).forEach(([key, value]) => {
		if (typeof value === "number") {
			if (!is32BitNumber(value)) logger.warn(`Pref key '${styleText("blue", key)}' is a number, but is more than 4 bytes, which can be problematic on some OS.`);
		}
	});
	if (dts) await outputFile(dts, renderPluginPrefsDts(prefsWithoutPrefix), "utf-8");
	if (prefixPrefKeys) {
		prefsManager.clearPrefs();
		prefsManager.setPrefs(prefsWithPrefix);
		await prefsManager.write(prefsFilePath);
	}
	if (prefixPrefKeys) {
		const HTML_PREFERENCE_PATTERN = /preference="(\S*)"/g;
		const xhtmlPaths = await glob(`${dist}/addon/**/*.xhtml`);
		await Promise.all(xhtmlPaths.map(async (path) => {
			let content = await readFile(path, "utf-8");
			const matchs = [...content.matchAll(HTML_PREFERENCE_PATTERN)];
			for (const match of matchs) {
				const [matched, key] = match;
				if (key.startsWith(prefix)) {
					logger.debug(`Pref key '${styleText("blue", key)}' is already starts with '${prefix}', skip prefixing it.`);
					continue;
				} else if (key.startsWith("extensions.")) {
					logger.warn(`Pref key '${styleText("blue", key)}' in ${styleText("gray", path)} starts with 'extensions.' but not '${styleText("blue", prefix)}', skip prefixing it.`);
					continue;
				} else if (!(key in prefsWithPrefix) && !(key in prefsWithoutPrefix)) {
					logger.warn(`Pref key '${styleText("blue", key)}' in ${styleText("gray", path)} is not found in prefs.js, skip prefixing it.`);
					continue;
				} else {
					const prefixed = `${prefix}.${key}`;
					logger.debug(`Pref key '${styleText("blue", key)}' in ${styleText("gray", path)} is prefixed to ${styleText("blue", prefixed)}.`);
					content = content.replace(matched, `preference="${prefixed}"`);
				}
			}
			await outputFile(path, content, "utf-8");
		}));
	}
}
function renderPluginPrefsDts(prefs) {
	return `// Generated by zotero-plugin-scaffold
/* prettier-ignore */
/* eslint-disable */
// @ts-nocheck

// prettier-ignore
declare namespace _ZoteroTypes {
  interface Prefs {
    PluginPrefsMap: {
      ${Object.entries(prefs).map(([key, value]) => {
		return `"${key}": ${typeof value};`;
	}).join("\n      ")}
    };
  }
}
`;
}
//#endregion
//#region src/core/builder/update-json.ts
function generateHash(filePath, algorithm) {
	return new Promise((resolve, reject) => {
		const hash = createHash(algorithm);
		const stream = createReadStream(filePath);
		stream.on("data", (data) => {
			hash.update(data);
		});
		stream.on("end", () => {
			resolve(`${algorithm}:${hash.digest("hex")}`);
		});
		stream.on("error", (error) => {
			reject(error);
		});
	});
}
async function buildUpdateJson(ctx) {
	const { dist, xpiName, id, version, xpiDownloadLink, build } = ctx;
	const manifest = await readJSON(`${dist}/addon/manifest.json`);
	const min = manifest.applications?.zotero?.strict_min_version;
	const max = manifest.applications?.zotero?.strict_max_version;
	const updateHash = await generateHash(`${dist}/${xpiName}.xpi`, "sha512");
	const data = { addons: { [id]: { updates: [...build.makeUpdateJson.updates, {
		version,
		update_link: xpiDownloadLink,
		...build.makeUpdateJson.hash && { update_hash: updateHash },
		applications: { zotero: {
			...min && { strict_min_version: min },
			...max && { strict_max_version: max }
		} }
	}] } } };
	await writeJson(`${dist}/update-beta.json`, data, { spaces: 2 });
	if (!ctx.templateData.isPreRelease) await writeJson(`${dist}/update.json`, data, { spaces: 2 });
	logger.debug(`Prepare Update.json for ${ctx.templateData.isPreRelease ? "\x1B[31m Prerelease \x1B[0m" : "\x1B[32m Release \x1B[0m"}`);
}
//#endregion
//#region src/core/builder/zip.ts
async function pack(dist, xpiName) {
	const zip = new AdmZip();
	zip.addLocalFolder(`${dist}/addon`);
	zip.writeZip(`${dist}/${xpiName}.xpi`);
}
//#endregion
//#region src/core/builder/index.ts
var Build = class extends Base {
	buildTime;
	constructor(ctx) {
		super(ctx);
		process$1.env.NODE_ENV ??= "production";
		this.buildTime = "";
	}
	/**
	* Default build runner
	*/
	async run() {
		const { dist, version } = this.ctx;
		const t = /* @__PURE__ */ new Date();
		this.buildTime = dateFormat("YYYY-mm-dd HH:MM:SS", t);
		this.logger.info(`Building version ${styleText("blue", version)} to ${styleText("blue", dist)} at ${styleText("blue", this.buildTime)} in ${styleText("blue", process$1.env.NODE_ENV)} mode.`);
		await this.ctx.hooks.callHook("build:init", this.ctx);
		this.logger.tip("Preparing static assets", { space: 1 });
		await this.prepareAssets();
		this.logger.tip("Bundling scripts", { space: 1 });
		await this.bundle();
		/** ======== build resolved =========== */
		if (process$1.env.NODE_ENV === "production") {
			this.logger.tip("Packing plugin", { space: 1 });
			await this.buildInProduction();
		}
		await this.ctx.hooks.callHook("build:done", this.ctx);
		this.logger.success(`Build finished in ${(Date.now() - t.getTime()) / 1e3} s.`);
	}
	async prepareAssets() {
		const { source, namespace, dist, build } = this.ctx;
		await emptyDir(dist);
		await this.ctx.hooks.callHook("build:mkdir", this.ctx);
		await copyAssets(source, dist, build.assets);
		await replaceDefine(dist, build.define);
		await this.ctx.hooks.callHook("build:copyAssets", this.ctx);
		this.logger.debug("Preparing manifest", { space: 2 });
		await buildManifest(this.ctx);
		await this.ctx.hooks.callHook("build:makeManifest", this.ctx);
		this.logger.debug("Preparing locale files", { space: 2 });
		await buildLocale(dist, namespace, build.fluent);
		await this.ctx.hooks.callHook("build:fluent", this.ctx);
		this.logger.debug("Preparing preference files", { space: 2 });
		await buildPrefs(dist, build.prefs);
	}
	async bundle() {
		const { dist, build: { esbuildOptions } } = this.ctx;
		await esbuild(dist, esbuildOptions);
		await this.ctx.hooks.callHook("build:bundle", this.ctx);
	}
	async buildInProduction() {
		const { dist, xpiName } = this.ctx;
		await pack(dist, xpiName);
		await this.ctx.hooks.callHook("build:pack", this.ctx);
		await buildUpdateJson(this.ctx);
		await this.ctx.hooks.callHook("build:makeUpdateJSON", this.ctx);
	}
	exit() {}
};
//#endregion
//#region src/core/releaser/base.ts
var ReleaseBase = class {
	isCI;
	ctx;
	constructor(ctx) {
		this.ctx = ctx;
		this.isCI = isCI;
	}
	checkFiles() {
		const { dist } = this.ctx;
		if (globSync(`${dist}/*.xpi`).length === 0) throw new Error("No xpi file found, are you sure you have run build?");
	}
	get logger() {
		return this.ctx.logger;
	}
};
//#endregion
//#region src/core/releaser/bump.ts
var Bump = class extends ReleaseBase {
	constructor(ctx) {
		super(ctx);
	}
	async run() {
		const bumppConfig = {
			...this.ctx.release.bumpp,
			push: true,
			progress: this.bumppProgress
		};
		const { version } = this.ctx;
		if (bumppConfig.release === version) {
			this.logger.debug("Commit, tag, and push are disabled because new version = old version.");
			bumppConfig.commit = false;
			bumppConfig.tag = false;
			bumppConfig.push = false;
		}
		const result = await versionBump(bumppConfig);
		this.ctx.version = result.newVersion;
		this.ctx.release.bumpp.tag = result.tag || this.ctx.release.bumpp.tag.toString().replace("%s", result.newVersion);
		this.ctx.release.bumpp.commit = result.commit || this.ctx.release.bumpp.commit.toString().replace("%s", result.newVersion);
		this.logger.debug(`The release context after bump: ", ${this.ctx.release}`);
	}
	/**
	* bumpp 显示进度的回调
	*
	* @see https://github.com/antfu/bumpp/blob/main/src/cli/index.ts
	*/
	get bumppProgress() {
		return ({ event, script, updatedFiles, skippedFiles, newVersion }) => {
			switch (event) {
				case ProgressEvent.FileUpdated:
					this.logger.success(`Updated ${updatedFiles.pop()} to ${newVersion}`);
					break;
				case ProgressEvent.FileSkipped:
					this.logger.info(`${skippedFiles.pop()} did not need to be updated`);
					break;
				case ProgressEvent.GitCommit:
					this.logger.success("Git commit");
					break;
				case ProgressEvent.GitTag:
					this.logger.success("Git tag");
					break;
				case ProgressEvent.GitPush:
					this.logger.success("Git push");
					break;
				case ProgressEvent.NpmScript:
					this.logger.success(`Npm run ${script}`);
					break;
			}
		};
	}
};
//#endregion
//#region node_modules/.pnpm/pathe@2.0.3/node_modules/pathe/dist/shared/pathe.M-eThtNZ.mjs
const _DRIVE_LETTER_START_RE = /^[A-Za-z]:\//;
function normalizeWindowsPath(input = "") {
	if (!input) return input;
	return input.replace(/\\/g, "/").replace(_DRIVE_LETTER_START_RE, (r) => r.toUpperCase());
}
const _UNC_REGEX = /^[/\\]{2}/;
const _IS_ABSOLUTE_RE = /^[/\\](?![/\\])|^[/\\]{2}(?!\.)|^[A-Za-z]:[/\\]/;
const _DRIVE_LETTER_RE = /^[A-Za-z]:$/;
const normalize = function(path) {
	if (path.length === 0) return ".";
	path = normalizeWindowsPath(path);
	const isUNCPath = path.match(_UNC_REGEX);
	const isPathAbsolute = isAbsolute$1(path);
	const trailingSeparator = path[path.length - 1] === "/";
	path = normalizeString(path, !isPathAbsolute);
	if (path.length === 0) {
		if (isPathAbsolute) return "/";
		return trailingSeparator ? "./" : ".";
	}
	if (trailingSeparator) path += "/";
	if (_DRIVE_LETTER_RE.test(path)) path += "/";
	if (isUNCPath) {
		if (!isPathAbsolute) return `//./${path}`;
		return `//${path}`;
	}
	return isPathAbsolute && !isAbsolute$1(path) ? `/${path}` : path;
};
const join$1 = function(...segments) {
	let path = "";
	for (const seg of segments) {
		if (!seg) continue;
		if (path.length > 0) {
			const pathTrailing = path[path.length - 1] === "/";
			const segLeading = seg[0] === "/";
			if (pathTrailing && segLeading) path += seg.slice(1);
			else path += pathTrailing || segLeading ? seg : `/${seg}`;
		} else path += seg;
	}
	return normalize(path);
};
function cwd() {
	if (typeof process !== "undefined" && typeof process.cwd === "function") return process.cwd().replace(/\\/g, "/");
	return "/";
}
const resolve$1 = function(...arguments_) {
	arguments_ = arguments_.map((argument) => normalizeWindowsPath(argument));
	let resolvedPath = "";
	let resolvedAbsolute = false;
	for (let index = arguments_.length - 1; index >= -1 && !resolvedAbsolute; index--) {
		const path = index >= 0 ? arguments_[index] : cwd();
		if (!path || path.length === 0) continue;
		resolvedPath = `${path}/${resolvedPath}`;
		resolvedAbsolute = isAbsolute$1(path);
	}
	resolvedPath = normalizeString(resolvedPath, !resolvedAbsolute);
	if (resolvedAbsolute && !isAbsolute$1(resolvedPath)) return `/${resolvedPath}`;
	return resolvedPath.length > 0 ? resolvedPath : ".";
};
function normalizeString(path, allowAboveRoot) {
	let res = "";
	let lastSegmentLength = 0;
	let lastSlash = -1;
	let dots = 0;
	let char = null;
	for (let index = 0; index <= path.length; ++index) {
		if (index < path.length) char = path[index];
		else if (char === "/") break;
		else char = "/";
		if (char === "/") {
			if (lastSlash === index - 1 || dots === 1);
			else if (dots === 2) {
				if (res.length < 2 || lastSegmentLength !== 2 || res[res.length - 1] !== "." || res[res.length - 2] !== ".") {
					if (res.length > 2) {
						const lastSlashIndex = res.lastIndexOf("/");
						if (lastSlashIndex === -1) {
							res = "";
							lastSegmentLength = 0;
						} else {
							res = res.slice(0, lastSlashIndex);
							lastSegmentLength = res.length - 1 - res.lastIndexOf("/");
						}
						lastSlash = index;
						dots = 0;
						continue;
					} else if (res.length > 0) {
						res = "";
						lastSegmentLength = 0;
						lastSlash = index;
						dots = 0;
						continue;
					}
				}
				if (allowAboveRoot) {
					res += res.length > 0 ? "/.." : "..";
					lastSegmentLength = 2;
				}
			} else {
				if (res.length > 0) res += `/${path.slice(lastSlash + 1, index)}`;
				else res = path.slice(lastSlash + 1, index);
				lastSegmentLength = index - lastSlash - 1;
			}
			lastSlash = index;
			dots = 0;
		} else if (char === "." && dots !== -1) ++dots;
		else dots = -1;
	}
	return res;
}
const isAbsolute$1 = function(p) {
	return _IS_ABSOLUTE_RE.test(p);
};
//#endregion
//#region node_modules/.pnpm/scule@1.3.0/node_modules/scule/dist/index.mjs
function upperFirst(str) {
	return str ? str[0].toUpperCase() + str.slice(1) : "";
}
//#endregion
//#region node_modules/.pnpm/convert-gitmoji@0.1.5/node_modules/convert-gitmoji/dist/index.mjs
const gitmojis = {
	":art:": "🎨",
	":zap:": "⚡️",
	":fire:": "🔥",
	":bug:": "🐛",
	":ambulance:": "🚑️",
	":sparkles:": "✨",
	":memo:": "📝",
	":rocket:": "🚀",
	":lipstick:": "💄",
	":tada:": "🎉",
	":white_check_mark:": "✅",
	":lock:": "🔒️",
	":closed_lock_with_key:": "🔐",
	":bookmark:": "🔖",
	":rotating_light:": "🚨",
	":construction:": "🚧",
	":green_heart:": "💚",
	":arrow_down:": "⬇️",
	":arrow_up:": "⬆️",
	":pushpin:": "📌",
	":construction_worker:": "👷",
	":chart_with_upwards_trend:": "📈",
	":recycle:": "♻️",
	":heavy_plus_sign:": "➕",
	":heavy_minus_sign:": "➖",
	":wrench:": "🔧",
	":hammer:": "🔨",
	":globe_with_meridians:": "🌐",
	":pencil2:": "✏️",
	":pencil:": "✏️",
	":poop:": "💩",
	":rewind:": "⏪️",
	":twisted_rightwards_arrows:": "🔀",
	":package:": "📦️",
	":alien:": "👽️",
	":truck:": "🚚",
	":page_facing_up:": "📄",
	":boom:": "💥",
	":bento:": "🍱",
	":wheelchair:": "♿️",
	":bulb:": "💡",
	":beers:": "🍻",
	":speech_balloon:": "💬",
	":card_file_box:": "🗃️",
	":loud_sound:": "🔊",
	":mute:": "🔇",
	":busts_in_silhouette:": "👥",
	":children_crossing:": "🚸",
	":building_construction:": "🏗️",
	":iphone:": "📱",
	":clown_face:": "🤡",
	":egg:": "🥚",
	":see_no_evil:": "🙈",
	":camera_flash:": "📸",
	":alembic:": "⚗️",
	":mag:": "🔍️",
	":label:": "🏷️",
	":seedling:": "🌱",
	":triangular_flag_on_post:": "🚩",
	":goal_net:": "🥅",
	":dizzy:": "💫",
	":wastebasket:": "🗑️",
	":passport_control:": "🛂",
	":adhesive_bandage:": "🩹",
	":monocle_face:": "🧐",
	":coffin:": "⚰️",
	":test_tube:": "🧪",
	":necktie:": "👔",
	":stethoscope:": "🩺",
	":bricks:": "🧱",
	":technologist:": "🧑‍💻",
	":money_with_wings:": "💸",
	":thread:": "🧵",
	":safety_vest:": "🦺"
};
function convert(content, withSpace) {
	const re = new RegExp(Object.keys(gitmojis).join("|"), "gi");
	return content.replace(re, function(matched) {
		switch (withSpace) {
			case true:
			case "trailing": return `${gitmojis[matched.toLowerCase()]} `;
			case "leading": return ` ${gitmojis[matched.toLowerCase()]}`;
			case "both": return ` ${gitmojis[matched.toLowerCase()]} `;
			default: return gitmojis[matched.toLowerCase()];
		}
	});
}
globalThis.Blob;
globalThis.File;
globalThis.FormData;
globalThis.Headers;
globalThis.Request;
globalThis.Response;
globalThis.AbortController;
//#endregion
//#region node_modules/.pnpm/node-fetch-native@1.6.7/node_modules/node-fetch-native/dist/native.mjs
const l = globalThis.fetch || (() => {
	throw new Error("[node-fetch-native] Failed to fetch: `globalThis.fetch` is not available!");
});
//#endregion
//#region node_modules/.pnpm/confbox@0.2.4/node_modules/confbox/dist/_chunks/_format.mjs
const t$1 = Symbol.for(`__confbox_fmt__`), n$2 = /^(\s+)/, r$2 = /(\s+)$/;
function i$1(e, t = {}) {
	return {
		sample: t.indent === void 0 && t.preserveIndentation !== !1 && e.slice(0, t?.sampleSize || 1024),
		whiteSpace: t.preserveWhitespace === !1 ? void 0 : {
			start: n$2.exec(e)?.[0] || ``,
			end: r$2.exec(e)?.[0] || ``
		}
	};
}
function a$1(e, n, r) {
	!n || typeof n != `object` || Object.defineProperty(n, t$1, {
		enumerable: !1,
		configurable: !0,
		writable: !0,
		value: i$1(e, r)
	});
}
//#endregion
//#region node_modules/.pnpm/confbox@0.2.4/node_modules/confbox/dist/_chunks/libs/jsonc-parser.mjs
function e(e, i = !1) {
	let a = e.length, o = 0, s = ``, c = 0, l = 16, u = 0, d = 0, f = 0, p = 0, m = 0;
	function h(t, n) {
		let r = 0, i = 0;
		for (; r < t || !n;) {
			let t = e.charCodeAt(o);
			if (t >= 48 && t <= 57) i = i * 16 + t - 48;
			else if (t >= 65 && t <= 70) i = i * 16 + t - 65 + 10;
			else if (t >= 97 && t <= 102) i = i * 16 + t - 97 + 10;
			else break;
			o++, r++;
		}
		return r < t && (i = -1), i;
	}
	function g(e) {
		o = e, s = ``, c = 0, l = 16, m = 0;
	}
	function _() {
		let t = o;
		if (e.charCodeAt(o) === 48) o++;
		else for (o++; o < e.length && r$1(e.charCodeAt(o));) o++;
		if (o < e.length && e.charCodeAt(o) === 46) if (o++, o < e.length && r$1(e.charCodeAt(o))) for (o++; o < e.length && r$1(e.charCodeAt(o));) o++;
		else return m = 3, e.substring(t, o);
		let n = o;
		if (o < e.length && (e.charCodeAt(o) === 69 || e.charCodeAt(o) === 101)) if (o++, (o < e.length && e.charCodeAt(o) === 43 || e.charCodeAt(o) === 45) && o++, o < e.length && r$1(e.charCodeAt(o))) {
			for (o++; o < e.length && r$1(e.charCodeAt(o));) o++;
			n = o;
		} else m = 3;
		return e.substring(t, n);
	}
	function v() {
		let t = ``, r = o;
		for (;;) {
			if (o >= a) {
				t += e.substring(r, o), m = 2;
				break;
			}
			let i = e.charCodeAt(o);
			if (i === 34) {
				t += e.substring(r, o), o++;
				break;
			}
			if (i === 92) {
				if (t += e.substring(r, o), o++, o >= a) {
					m = 2;
					break;
				}
				switch (e.charCodeAt(o++)) {
					case 34:
						t += `"`;
						break;
					case 92:
						t += `\\`;
						break;
					case 47:
						t += `/`;
						break;
					case 98:
						t += `\b`;
						break;
					case 102:
						t += `\f`;
						break;
					case 110:
						t += `
`;
						break;
					case 114:
						t += `\r`;
						break;
					case 116:
						t += `	`;
						break;
					case 117:
						let e = h(4, !0);
						e >= 0 ? t += String.fromCharCode(e) : m = 4;
						break;
					default: m = 5;
				}
				r = o;
				continue;
			}
			if (i >= 0 && i <= 31) if (n$1(i)) {
				t += e.substring(r, o), m = 2;
				break;
			} else m = 6;
			o++;
		}
		return t;
	}
	function y() {
		if (s = ``, m = 0, c = o, d = u, p = f, o >= a) return c = a, l = 17;
		let i = e.charCodeAt(o);
		if (t(i)) {
			do
				o++, s += String.fromCharCode(i), i = e.charCodeAt(o);
			while (t(i));
			return l = 15;
		}
		if (n$1(i)) return o++, s += String.fromCharCode(i), i === 13 && e.charCodeAt(o) === 10 && (o++, s += `
`), u++, f = o, l = 14;
		switch (i) {
			case 123: return o++, l = 1;
			case 125: return o++, l = 2;
			case 91: return o++, l = 3;
			case 93: return o++, l = 4;
			case 58: return o++, l = 6;
			case 44: return o++, l = 5;
			case 34: return o++, s = v(), l = 10;
			case 47:
				let t = o - 1;
				if (e.charCodeAt(o + 1) === 47) {
					for (o += 2; o < a && !n$1(e.charCodeAt(o));) o++;
					return s = e.substring(t, o), l = 12;
				}
				if (e.charCodeAt(o + 1) === 42) {
					o += 2;
					let r = a - 1, i = !1;
					for (; o < r;) {
						let t = e.charCodeAt(o);
						if (t === 42 && e.charCodeAt(o + 1) === 47) {
							o += 2, i = !0;
							break;
						}
						o++, n$1(t) && (t === 13 && e.charCodeAt(o) === 10 && o++, u++, f = o);
					}
					return i || (o++, m = 1), s = e.substring(t, o), l = 13;
				}
				return s += String.fromCharCode(i), o++, l = 16;
			case 45: if (s += String.fromCharCode(i), o++, o === a || !r$1(e.charCodeAt(o))) return l = 16;
			case 48:
			case 49:
			case 50:
			case 51:
			case 52:
			case 53:
			case 54:
			case 55:
			case 56:
			case 57: return s += _(), l = 11;
			default:
				for (; o < a && b(i);) o++, i = e.charCodeAt(o);
				if (c !== o) {
					switch (s = e.substring(c, o), s) {
						case `true`: return l = 8;
						case `false`: return l = 9;
						case `null`: return l = 7;
					}
					return l = 16;
				}
				return s += String.fromCharCode(i), o++, l = 16;
		}
	}
	function b(e) {
		if (t(e) || n$1(e)) return !1;
		switch (e) {
			case 125:
			case 93:
			case 123:
			case 91:
			case 34:
			case 58:
			case 44:
			case 47: return !1;
		}
		return !0;
	}
	function x() {
		let e;
		do
			e = y();
		while (e >= 12 && e <= 15);
		return e;
	}
	return {
		setPosition: g,
		getPosition: () => o,
		scan: i ? x : y,
		getToken: () => l,
		getTokenValue: () => s,
		getTokenOffset: () => c,
		getTokenLength: () => o - c,
		getTokenStartLine: () => d,
		getTokenStartCharacter: () => c - p,
		getTokenError: () => m
	};
}
function t(e) {
	return e === 32 || e === 9;
}
function n$1(e) {
	return e === 10 || e === 13;
}
function r$1(e) {
	return e >= 48 && e <= 57;
}
var i;
(function(e) {
	e[e.lineFeed = 10] = `lineFeed`, e[e.carriageReturn = 13] = `carriageReturn`, e[e.space = 32] = `space`, e[e._0 = 48] = `_0`, e[e._1 = 49] = `_1`, e[e._2 = 50] = `_2`, e[e._3 = 51] = `_3`, e[e._4 = 52] = `_4`, e[e._5 = 53] = `_5`, e[e._6 = 54] = `_6`, e[e._7 = 55] = `_7`, e[e._8 = 56] = `_8`, e[e._9 = 57] = `_9`, e[e.a = 97] = `a`, e[e.b = 98] = `b`, e[e.c = 99] = `c`, e[e.d = 100] = `d`, e[e.e = 101] = `e`, e[e.f = 102] = `f`, e[e.g = 103] = `g`, e[e.h = 104] = `h`, e[e.i = 105] = `i`, e[e.j = 106] = `j`, e[e.k = 107] = `k`, e[e.l = 108] = `l`, e[e.m = 109] = `m`, e[e.n = 110] = `n`, e[e.o = 111] = `o`, e[e.p = 112] = `p`, e[e.q = 113] = `q`, e[e.r = 114] = `r`, e[e.s = 115] = `s`, e[e.t = 116] = `t`, e[e.u = 117] = `u`, e[e.v = 118] = `v`, e[e.w = 119] = `w`, e[e.x = 120] = `x`, e[e.y = 121] = `y`, e[e.z = 122] = `z`, e[e.A = 65] = `A`, e[e.B = 66] = `B`, e[e.C = 67] = `C`, e[e.D = 68] = `D`, e[e.E = 69] = `E`, e[e.F = 70] = `F`, e[e.G = 71] = `G`, e[e.H = 72] = `H`, e[e.I = 73] = `I`, e[e.J = 74] = `J`, e[e.K = 75] = `K`, e[e.L = 76] = `L`, e[e.M = 77] = `M`, e[e.N = 78] = `N`, e[e.O = 79] = `O`, e[e.P = 80] = `P`, e[e.Q = 81] = `Q`, e[e.R = 82] = `R`, e[e.S = 83] = `S`, e[e.T = 84] = `T`, e[e.U = 85] = `U`, e[e.V = 86] = `V`, e[e.W = 87] = `W`, e[e.X = 88] = `X`, e[e.Y = 89] = `Y`, e[e.Z = 90] = `Z`, e[e.asterisk = 42] = `asterisk`, e[e.backslash = 92] = `backslash`, e[e.closeBrace = 125] = `closeBrace`, e[e.closeBracket = 93] = `closeBracket`, e[e.colon = 58] = `colon`, e[e.comma = 44] = `comma`, e[e.dot = 46] = `dot`, e[e.doubleQuote = 34] = `doubleQuote`, e[e.minus = 45] = `minus`, e[e.openBrace = 123] = `openBrace`, e[e.openBracket = 91] = `openBracket`, e[e.plus = 43] = `plus`, e[e.slash = 47] = `slash`, e[e.formFeed = 12] = `formFeed`, e[e.tab = 9] = `tab`;
})(i ||= {}), Array(20).fill(0).map((e, t) => ` `.repeat(t)), Array(200).fill(0).map((e, t) => `
` + ` `.repeat(t)), Array(200).fill(0).map((e, t) => `\r` + ` `.repeat(t)), Array(200).fill(0).map((e, t) => `\r
` + ` `.repeat(t)), Array(200).fill(0).map((e, t) => `
` + `	`.repeat(t)), Array(200).fill(0).map((e, t) => `\r` + `	`.repeat(t)), Array(200).fill(0).map((e, t) => `\r
` + `	`.repeat(t));
var a;
(function(e) {
	e.DEFAULT = { allowTrailingComma: !1 };
})(a ||= {});
function o(e, t = [], n = a.DEFAULT) {
	let r = null, i = [], o = [];
	function s(e) {
		Array.isArray(i) ? i.push(e) : r !== null && (i[r] = e);
	}
	return d(e, {
		onObjectBegin: () => {
			let e = {};
			s(e), o.push(i), i = e, r = null;
		},
		onObjectProperty: (e) => {
			r = e;
		},
		onObjectEnd: () => {
			i = o.pop();
		},
		onArrayBegin: () => {
			let e = [];
			s(e), o.push(i), i = e, r = null;
		},
		onArrayEnd: () => {
			i = o.pop();
		},
		onLiteralValue: s,
		onError: (e, n, r) => {
			t.push({
				error: e,
				offset: n,
				length: r
			});
		}
	}, n), i[0];
}
function d(t, n, r = a.DEFAULT) {
	let i = e(t, !1), o = [], s = 0;
	function c(e) {
		return e ? () => s === 0 && e(i.getTokenOffset(), i.getTokenLength(), i.getTokenStartLine(), i.getTokenStartCharacter()) : () => !0;
	}
	function l(e) {
		return e ? (t) => s === 0 && e(t, i.getTokenOffset(), i.getTokenLength(), i.getTokenStartLine(), i.getTokenStartCharacter()) : () => !0;
	}
	function u(e) {
		return e ? (t) => s === 0 && e(t, i.getTokenOffset(), i.getTokenLength(), i.getTokenStartLine(), i.getTokenStartCharacter(), () => o.slice()) : () => !0;
	}
	function d(e) {
		return e ? () => {
			s > 0 ? s++ : e(i.getTokenOffset(), i.getTokenLength(), i.getTokenStartLine(), i.getTokenStartCharacter(), () => o.slice()) === !1 && (s = 1);
		} : () => !0;
	}
	function f(e) {
		return e ? () => {
			s > 0 && s--, s === 0 && e(i.getTokenOffset(), i.getTokenLength(), i.getTokenStartLine(), i.getTokenStartCharacter());
		} : () => !0;
	}
	let p = d(n.onObjectBegin), m = u(n.onObjectProperty), h = f(n.onObjectEnd), g = d(n.onArrayBegin), _ = f(n.onArrayEnd), v = u(n.onLiteralValue), y = l(n.onSeparator), b = c(n.onComment), x = l(n.onError), S = r && r.disallowComments, C = r && r.allowTrailingComma;
	function w() {
		for (;;) {
			let e = i.scan();
			switch (i.getTokenError()) {
				case 4:
					T(14);
					break;
				case 5:
					T(15);
					break;
				case 3:
					T(13);
					break;
				case 1:
					S || T(11);
					break;
				case 2:
					T(12);
					break;
				case 6:
					T(16);
					break;
			}
			switch (e) {
				case 12:
				case 13:
					S ? T(10) : b();
					break;
				case 16:
					T(1);
					break;
				case 15:
				case 14: break;
				default: return e;
			}
		}
	}
	function T(e, t = [], n = []) {
		if (x(e), t.length + n.length > 0) {
			let e = i.getToken();
			for (; e !== 17;) {
				if (t.indexOf(e) !== -1) {
					w();
					break;
				} else if (n.indexOf(e) !== -1) break;
				e = w();
			}
		}
	}
	function E(e) {
		let t = i.getTokenValue();
		return e ? v(t) : (m(t), o.push(t)), w(), !0;
	}
	function D() {
		switch (i.getToken()) {
			case 11:
				let e = i.getTokenValue(), t = Number(e);
				isNaN(t) && (T(2), t = 0), v(t);
				break;
			case 7:
				v(null);
				break;
			case 8:
				v(!0);
				break;
			case 9:
				v(!1);
				break;
			default: return !1;
		}
		return w(), !0;
	}
	function O() {
		return i.getToken() === 10 ? (E(!1), i.getToken() === 6 ? (y(`:`), w(), j() || T(4, [], [2, 5])) : T(5, [], [2, 5]), o.pop(), !0) : (T(3, [], [2, 5]), !1);
	}
	function k() {
		p(), w();
		let e = !1;
		for (; i.getToken() !== 2 && i.getToken() !== 17;) {
			if (i.getToken() === 5) {
				if (e || T(4, [], []), y(`,`), w(), i.getToken() === 2 && C) break;
			} else e && T(6, [], []);
			O() || T(4, [], [2, 5]), e = !0;
		}
		return h(), i.getToken() === 2 ? w() : T(7, [2], []), !0;
	}
	function A() {
		g(), w();
		let e = !0, t = !1;
		for (; i.getToken() !== 4 && i.getToken() !== 17;) {
			if (i.getToken() === 5) {
				if (t || T(4, [], []), y(`,`), w(), i.getToken() === 4 && C) break;
			} else t && T(6, [], []);
			e ? (o.push(0), e = !1) : o[o.length - 1]++, j() || T(4, [], [4, 5]), t = !0;
		}
		return _(), e || o.pop(), i.getToken() === 4 ? w() : T(8, [4], []), !0;
	}
	function j() {
		switch (i.getToken()) {
			case 3: return A();
			case 1: return k();
			case 10: return E(!0);
			default: return D();
		}
	}
	return w(), i.getToken() === 17 ? r.allowEmptyContent ? !0 : (T(4, [], []), !1) : j() ? (i.getToken() !== 17 && T(9, [], []), !0) : (T(4, [], []), !1);
}
var f;
(function(e) {
	e[e.None = 0] = `None`, e[e.UnexpectedEndOfComment = 1] = `UnexpectedEndOfComment`, e[e.UnexpectedEndOfString = 2] = `UnexpectedEndOfString`, e[e.UnexpectedEndOfNumber = 3] = `UnexpectedEndOfNumber`, e[e.InvalidUnicode = 4] = `InvalidUnicode`, e[e.InvalidEscapeCharacter = 5] = `InvalidEscapeCharacter`, e[e.InvalidCharacter = 6] = `InvalidCharacter`;
})(f ||= {});
var p;
(function(e) {
	e[e.OpenBraceToken = 1] = `OpenBraceToken`, e[e.CloseBraceToken = 2] = `CloseBraceToken`, e[e.OpenBracketToken = 3] = `OpenBracketToken`, e[e.CloseBracketToken = 4] = `CloseBracketToken`, e[e.CommaToken = 5] = `CommaToken`, e[e.ColonToken = 6] = `ColonToken`, e[e.NullKeyword = 7] = `NullKeyword`, e[e.TrueKeyword = 8] = `TrueKeyword`, e[e.FalseKeyword = 9] = `FalseKeyword`, e[e.StringLiteral = 10] = `StringLiteral`, e[e.NumericLiteral = 11] = `NumericLiteral`, e[e.LineCommentTrivia = 12] = `LineCommentTrivia`, e[e.BlockCommentTrivia = 13] = `BlockCommentTrivia`, e[e.LineBreakTrivia = 14] = `LineBreakTrivia`, e[e.Trivia = 15] = `Trivia`, e[e.Unknown = 16] = `Unknown`, e[e.EOF = 17] = `EOF`;
})(p ||= {});
const m = o;
var h;
(function(e) {
	e[e.InvalidSymbol = 1] = `InvalidSymbol`, e[e.InvalidNumberFormat = 2] = `InvalidNumberFormat`, e[e.PropertyNameExpected = 3] = `PropertyNameExpected`, e[e.ValueExpected = 4] = `ValueExpected`, e[e.ColonExpected = 5] = `ColonExpected`, e[e.CommaExpected = 6] = `CommaExpected`, e[e.CloseBraceExpected = 7] = `CloseBraceExpected`, e[e.CloseBracketExpected = 8] = `CloseBracketExpected`, e[e.EndOfFileExpected = 9] = `EndOfFileExpected`, e[e.InvalidCommentToken = 10] = `InvalidCommentToken`, e[e.UnexpectedEndOfComment = 11] = `UnexpectedEndOfComment`, e[e.UnexpectedEndOfString = 12] = `UnexpectedEndOfString`, e[e.UnexpectedEndOfNumber = 13] = `UnexpectedEndOfNumber`, e[e.InvalidUnicode = 14] = `InvalidUnicode`, e[e.InvalidEscapeCharacter = 15] = `InvalidEscapeCharacter`, e[e.InvalidCharacter = 16] = `InvalidCharacter`;
})(h ||= {});
//#endregion
//#region node_modules/.pnpm/confbox@0.2.4/node_modules/confbox/dist/_chunks/json.mjs
function n(t, n) {
	let r = JSON.parse(t, n?.reviver);
	return a$1(t, r, n), r;
}
//#endregion
//#region node_modules/.pnpm/confbox@0.2.4/node_modules/confbox/dist/jsonc.mjs
function r(n, r) {
	let i = m(n, r?.errors, r);
	return a$1(n, i, r), i;
}
//#endregion
//#region node_modules/.pnpm/exsolve@1.0.8/node_modules/exsolve/dist/index.mjs
const nodeBuiltins = [
	"_http_agent",
	"_http_client",
	"_http_common",
	"_http_incoming",
	"_http_outgoing",
	"_http_server",
	"_stream_duplex",
	"_stream_passthrough",
	"_stream_readable",
	"_stream_transform",
	"_stream_wrap",
	"_stream_writable",
	"_tls_common",
	"_tls_wrap",
	"assert",
	"assert/strict",
	"async_hooks",
	"buffer",
	"child_process",
	"cluster",
	"console",
	"constants",
	"crypto",
	"dgram",
	"diagnostics_channel",
	"dns",
	"dns/promises",
	"domain",
	"events",
	"fs",
	"fs/promises",
	"http",
	"http2",
	"https",
	"inspector",
	"inspector/promises",
	"module",
	"net",
	"os",
	"path",
	"path/posix",
	"path/win32",
	"perf_hooks",
	"process",
	"punycode",
	"querystring",
	"readline",
	"readline/promises",
	"repl",
	"stream",
	"stream/consumers",
	"stream/promises",
	"stream/web",
	"string_decoder",
	"sys",
	"timers",
	"timers/promises",
	"tls",
	"trace_events",
	"tty",
	"url",
	"util",
	"util/types",
	"v8",
	"vm",
	"wasi",
	"worker_threads",
	"zlib"
];
const own$1 = {}.hasOwnProperty;
const classRegExp = /^([A-Z][a-z\d]*)+$/;
const kTypes = new Set([
	"string",
	"function",
	"number",
	"object",
	"Function",
	"Object",
	"boolean",
	"bigint",
	"symbol"
]);
const messages = /* @__PURE__ */ new Map();
const nodeInternalPrefix = "__node_internal_";
let userStackTraceLimit;
/**
* Create a list string in the form like 'A and B' or 'A, B, ..., and Z'.
* We cannot use Intl.ListFormat because it's not available in
* --without-intl builds.
*
* @param {Array<string>} array
*   An array of strings.
* @param {string} [type]
*   The list type to be inserted before the last element.
* @returns {string}
*/
function formatList(array, type = "and") {
	return array.length < 3 ? array.join(` ${type} `) : `${array.slice(0, -1).join(", ")}, ${type} ${array.at(-1)}`;
}
/**
* Utility function for registering the error codes.
*/
function createError(sym, value, constructor) {
	messages.set(sym, value);
	return makeNodeErrorWithCode(constructor, sym);
}
function makeNodeErrorWithCode(Base, key) {
	return function NodeError(...parameters) {
		const limit = Error.stackTraceLimit;
		if (isErrorStackTraceLimitWritable()) Error.stackTraceLimit = 0;
		const error = new Base();
		if (isErrorStackTraceLimitWritable()) Error.stackTraceLimit = limit;
		const message = getMessage(key, parameters, error);
		Object.defineProperties(error, {
			message: {
				value: message,
				enumerable: false,
				writable: true,
				configurable: true
			},
			toString: {
				value() {
					return `${this.name} [${key}]: ${this.message}`;
				},
				enumerable: false,
				writable: true,
				configurable: true
			}
		});
		captureLargerStackTrace(error);
		error.code = key;
		return error;
	};
}
function isErrorStackTraceLimitWritable() {
	try {
		if (v8.startupSnapshot.isBuildingSnapshot()) return false;
	} catch {}
	const desc = Object.getOwnPropertyDescriptor(Error, "stackTraceLimit");
	if (desc === void 0) return Object.isExtensible(Error);
	return own$1.call(desc, "writable") && desc.writable !== void 0 ? desc.writable : desc.set !== void 0;
}
/**
* This function removes unnecessary frames from Node.js core errors.
*/
function hideStackFrames(wrappedFunction) {
	const hidden = nodeInternalPrefix + wrappedFunction.name;
	Object.defineProperty(wrappedFunction, "name", { value: hidden });
	return wrappedFunction;
}
const captureLargerStackTrace = hideStackFrames(function(error) {
	const stackTraceLimitIsWritable = isErrorStackTraceLimitWritable();
	if (stackTraceLimitIsWritable) {
		userStackTraceLimit = Error.stackTraceLimit;
		Error.stackTraceLimit = Number.POSITIVE_INFINITY;
	}
	Error.captureStackTrace(error);
	if (stackTraceLimitIsWritable) Error.stackTraceLimit = userStackTraceLimit;
	return error;
});
function getMessage(key, parameters, self) {
	const message = messages.get(key);
	assert.ok(message !== void 0, "expected `message` to be found");
	if (typeof message === "function") {
		assert.ok(message.length <= parameters.length, `Code: ${key}; The provided arguments length (${parameters.length}) does not match the required ones (${message.length}).`);
		return Reflect.apply(message, self, parameters);
	}
	const regex = /%[dfijoOs]/g;
	let expectedLength = 0;
	while (regex.exec(message) !== null) expectedLength++;
	assert.ok(expectedLength === parameters.length, `Code: ${key}; The provided arguments length (${parameters.length}) does not match the required ones (${expectedLength}).`);
	if (parameters.length === 0) return message;
	parameters.unshift(message);
	return Reflect.apply(format, null, parameters);
}
/**
* Determine the specific type of a value for type-mismatch errors.
*/
function determineSpecificType(value) {
	if (value === null || value === void 0) return String(value);
	if (typeof value === "function" && value.name) return `function ${value.name}`;
	if (typeof value === "object") {
		if (value.constructor && value.constructor.name) return `an instance of ${value.constructor.name}`;
		return `${inspect(value, { depth: -1 })}`;
	}
	let inspected = inspect(value, { colors: false });
	if (inspected.length > 28) inspected = `${inspected.slice(0, 25)}...`;
	return `type ${typeof value} (${inspected})`;
}
createError("ERR_INVALID_ARG_TYPE", (name, expected, actual) => {
	assert.ok(typeof name === "string", "'name' must be a string");
	if (!Array.isArray(expected)) expected = [expected];
	let message = "The ";
	if (name.endsWith(" argument")) message += `${name} `;
	else {
		const type = name.includes(".") ? "property" : "argument";
		message += `"${name}" ${type} `;
	}
	message += "must be ";
	const types = [];
	const instances = [];
	const other = [];
	for (const value of expected) {
		assert.ok(typeof value === "string", "All expected entries have to be of type string");
		if (kTypes.has(value)) types.push(value.toLowerCase());
		else if (classRegExp.exec(value) === null) {
			assert.ok(value !== "object", "The value \"object\" should be written as \"Object\"");
			other.push(value);
		} else instances.push(value);
	}
	if (instances.length > 0) {
		const pos = types.indexOf("object");
		if (pos !== -1) {
			types.slice(pos, 1);
			instances.push("Object");
		}
	}
	if (types.length > 0) {
		message += `${types.length > 1 ? "one of type" : "of type"} ${formatList(types, "or")}`;
		if (instances.length > 0 || other.length > 0) message += " or ";
	}
	if (instances.length > 0) {
		message += `an instance of ${formatList(instances, "or")}`;
		if (other.length > 0) message += " or ";
	}
	if (other.length > 0) if (other.length > 1) message += `one of ${formatList(other, "or")}`;
	else {
		if (other[0]?.toLowerCase() !== other[0]) message += "an ";
		message += `${other[0]}`;
	}
	message += `. Received ${determineSpecificType(actual)}`;
	return message;
}, TypeError);
const ERR_INVALID_MODULE_SPECIFIER = createError(
	"ERR_INVALID_MODULE_SPECIFIER",
	/**
	* @param {string} request
	* @param {string} reason
	* @param {string} [base]
	*/
	(request, reason, base) => {
		return `Invalid module "${request}" ${reason}${base ? ` imported from ${base}` : ""}`;
	},
	TypeError
);
const ERR_INVALID_PACKAGE_CONFIG = createError("ERR_INVALID_PACKAGE_CONFIG", (path$1, base, message) => {
	return `Invalid package config ${path$1}${base ? ` while importing ${base}` : ""}${message ? `. ${message}` : ""}`;
}, Error);
const ERR_INVALID_PACKAGE_TARGET = createError("ERR_INVALID_PACKAGE_TARGET", (packagePath, key, target, isImport = false, base) => {
	const relatedError = typeof target === "string" && !isImport && target.length > 0 && !target.startsWith("./");
	if (key === ".") {
		assert.ok(isImport === false);
		return `Invalid "exports" main target ${JSON.stringify(target)} defined in the package config ${packagePath}package.json${base ? ` imported from ${base}` : ""}${relatedError ? "; targets must start with \"./\"" : ""}`;
	}
	return `Invalid "${isImport ? "imports" : "exports"}" target ${JSON.stringify(target)} defined for '${key}' in the package config ${packagePath}package.json${base ? ` imported from ${base}` : ""}${relatedError ? "; targets must start with \"./\"" : ""}`;
}, Error);
const ERR_MODULE_NOT_FOUND = createError("ERR_MODULE_NOT_FOUND", (path$1, base, exactUrl = false) => {
	return `Cannot find ${exactUrl ? "module" : "package"} '${path$1}' imported from ${base}`;
}, Error);
createError("ERR_NETWORK_IMPORT_DISALLOWED", "import of '%s' by %s is not supported: %s", Error);
const ERR_PACKAGE_IMPORT_NOT_DEFINED = createError("ERR_PACKAGE_IMPORT_NOT_DEFINED", (specifier, packagePath, base) => {
	return `Package import specifier "${specifier}" is not defined${packagePath ? ` in package ${packagePath || ""}package.json` : ""} imported from ${base}`;
}, TypeError);
const ERR_PACKAGE_PATH_NOT_EXPORTED = createError(
	"ERR_PACKAGE_PATH_NOT_EXPORTED",
	/**
	* @param {string} packagePath
	* @param {string} subpath
	* @param {string} [base]
	*/
	(packagePath, subpath, base) => {
		if (subpath === ".") return `No "exports" main defined in ${packagePath}package.json${base ? ` imported from ${base}` : ""}`;
		return `Package subpath '${subpath}' is not defined by "exports" in ${packagePath}package.json${base ? ` imported from ${base}` : ""}`;
	},
	Error
);
const ERR_UNSUPPORTED_DIR_IMPORT = createError("ERR_UNSUPPORTED_DIR_IMPORT", "Directory import '%s' is not supported resolving ES modules imported from %s", Error);
const ERR_UNSUPPORTED_RESOLVE_REQUEST = createError("ERR_UNSUPPORTED_RESOLVE_REQUEST", "Failed to resolve module specifier \"%s\" from \"%s\": Invalid relative URL or base scheme is not hierarchical.", TypeError);
const ERR_UNKNOWN_FILE_EXTENSION = createError("ERR_UNKNOWN_FILE_EXTENSION", (extension, path$1) => {
	return `Unknown file extension "${extension}" for ${path$1}`;
}, TypeError);
createError("ERR_INVALID_ARG_VALUE", (name, value, reason = "is invalid") => {
	let inspected = inspect(value);
	if (inspected.length > 128) inspected = `${inspected.slice(0, 128)}...`;
	return `The ${name.includes(".") ? "property" : "argument"} '${name}' ${reason}. Received ${inspected}`;
}, TypeError);
const hasOwnProperty$1 = {}.hasOwnProperty;
const cache = /* @__PURE__ */ new Map();
function read(jsonPath, { base, specifier }) {
	const existing = cache.get(jsonPath);
	if (existing) return existing;
	let string;
	try {
		string = fs.readFileSync(path.toNamespacedPath(jsonPath), "utf8");
	} catch (error) {
		const exception = error;
		if (exception.code !== "ENOENT") throw exception;
	}
	const result = {
		exists: false,
		pjsonPath: jsonPath,
		main: void 0,
		name: void 0,
		type: "none",
		exports: void 0,
		imports: void 0
	};
	if (string !== void 0) {
		let parsed;
		try {
			parsed = JSON.parse(string);
		} catch (error_) {
			const error = new ERR_INVALID_PACKAGE_CONFIG(jsonPath, (base ? `"${specifier}" from ` : "") + fileURLToPath(base || specifier), error_.message);
			error.cause = error_;
			throw error;
		}
		result.exists = true;
		if (hasOwnProperty$1.call(parsed, "name") && typeof parsed.name === "string") result.name = parsed.name;
		if (hasOwnProperty$1.call(parsed, "main") && typeof parsed.main === "string") result.main = parsed.main;
		if (hasOwnProperty$1.call(parsed, "exports")) result.exports = parsed.exports;
		if (hasOwnProperty$1.call(parsed, "imports")) result.imports = parsed.imports;
		if (hasOwnProperty$1.call(parsed, "type") && (parsed.type === "commonjs" || parsed.type === "module")) result.type = parsed.type;
	}
	cache.set(jsonPath, result);
	return result;
}
function getPackageScopeConfig(resolved) {
	let packageJSONUrl = new URL("package.json", resolved);
	while (true) {
		if (packageJSONUrl.pathname.endsWith("node_modules/package.json")) break;
		const packageConfig = read(fileURLToPath(packageJSONUrl), { specifier: resolved });
		if (packageConfig.exists) return packageConfig;
		const lastPackageJSONUrl = packageJSONUrl;
		packageJSONUrl = new URL("../package.json", packageJSONUrl);
		if (packageJSONUrl.pathname === lastPackageJSONUrl.pathname) break;
	}
	return {
		pjsonPath: fileURLToPath(packageJSONUrl),
		exists: false,
		type: "none"
	};
}
const hasOwnProperty = {}.hasOwnProperty;
const extensionFormatMap = {
	__proto__: null,
	".json": "json",
	".cjs": "commonjs",
	".cts": "commonjs",
	".js": "module",
	".ts": "module",
	".mts": "module",
	".mjs": "module"
};
const protocolHandlers = {
	__proto__: null,
	"data:": getDataProtocolModuleFormat,
	"file:": getFileProtocolModuleFormat,
	"node:": () => "builtin"
};
function mimeToFormat(mime) {
	if (mime && /\s*(text|application)\/javascript\s*(;\s*charset=utf-?8\s*)?/i.test(mime)) return "module";
	if (mime === "application/json") return "json";
	return null;
}
function getDataProtocolModuleFormat(parsed) {
	const { 1: mime } = /^([^/]+\/[^;,]+)[^,]*?(;base64)?,/.exec(parsed.pathname) || [
		null,
		null,
		null
	];
	return mimeToFormat(mime);
}
/**
* Returns the file extension from a URL.
*
* Should give similar result to
* `require('node:path').extname(require('node:url').fileURLToPath(url))`
* when used with a `file:` URL.
*
*/
function extname$1(url) {
	const pathname = url.pathname;
	let index = pathname.length;
	while (index--) {
		const code = pathname.codePointAt(index);
		if (code === 47) return "";
		if (code === 46) return pathname.codePointAt(index - 1) === 47 ? "" : pathname.slice(index);
	}
	return "";
}
function getFileProtocolModuleFormat(url, _context, ignoreErrors) {
	const ext = extname$1(url);
	if (ext === ".js") {
		const { type: packageType } = getPackageScopeConfig(url);
		if (packageType !== "none") return packageType;
		return "commonjs";
	}
	if (ext === "") {
		const { type: packageType } = getPackageScopeConfig(url);
		if (packageType === "none" || packageType === "commonjs") return "commonjs";
		return "module";
	}
	const format$1 = extensionFormatMap[ext];
	if (format$1) return format$1;
	if (ignoreErrors) return;
	throw new ERR_UNKNOWN_FILE_EXTENSION(ext, fileURLToPath(url));
}
function defaultGetFormatWithoutErrors(url, context) {
	const protocol = url.protocol;
	if (!hasOwnProperty.call(protocolHandlers, protocol)) return null;
	return protocolHandlers[protocol](url, context, true) || null;
}
const RegExpPrototypeSymbolReplace = RegExp.prototype[Symbol.replace];
const own = {}.hasOwnProperty;
const invalidSegmentRegEx = /(^|\\|\/)((\.|%2e)(\.|%2e)?|(n|%6e|%4e)(o|%6f|%4f)(d|%64|%44)(e|%65|%45)(_|%5f)(m|%6d|%4d)(o|%6f|%4f)(d|%64|%44)(u|%75|%55)(l|%6c|%4c)(e|%65|%45)(s|%73|%53))?(\\|\/|$)/i;
const deprecatedInvalidSegmentRegEx = /(^|\\|\/)((\.|%2e)(\.|%2e)?|(n|%6e|%4e)(o|%6f|%4f)(d|%64|%44)(e|%65|%45)(_|%5f)(m|%6d|%4d)(o|%6f|%4f)(d|%64|%44)(u|%75|%55)(l|%6c|%4c)(e|%65|%45)(s|%73|%53))(\\|\/|$)/i;
const invalidPackageNameRegEx = /^\.|%|\\/;
const patternRegEx = /\*/g;
const encodedSeparatorRegEx = /%2f|%5c/i;
const emittedPackageWarnings = /* @__PURE__ */ new Set();
const doubleSlashRegEx = /[/\\]{2}/;
function emitInvalidSegmentDeprecation(target, request, match, packageJsonUrl, internal, base, isTarget) {
	if (process$1.noDeprecation) return;
	const pjsonPath = fileURLToPath(packageJsonUrl);
	const double = doubleSlashRegEx.exec(isTarget ? target : request) !== null;
	process$1.emitWarning(`Use of deprecated ${double ? "double slash" : "leading or trailing slash matching"} resolving "${target}" for module request "${request}" ${request === match ? "" : `matched to "${match}" `}in the "${internal ? "imports" : "exports"}" field module resolution of the package at ${pjsonPath}${base ? ` imported from ${fileURLToPath(base)}` : ""}.`, "DeprecationWarning", "DEP0166");
}
function emitLegacyIndexDeprecation(url, packageJsonUrl, base, main) {
	if (process$1.noDeprecation) return;
	if (defaultGetFormatWithoutErrors(url, { parentURL: base.href }) !== "module") return;
	const urlPath = fileURLToPath(url.href);
	const packagePath = fileURLToPath(new URL$1(".", packageJsonUrl));
	const basePath = fileURLToPath(base);
	if (!main) process$1.emitWarning(`No "main" or "exports" field defined in the package.json for ${packagePath} resolving the main entry point "${urlPath.slice(packagePath.length)}", imported from ${basePath}.\nDefault "index" lookups for the main are deprecated for ES modules.`, "DeprecationWarning", "DEP0151");
	else if (path.resolve(packagePath, main) !== urlPath) process$1.emitWarning(`Package ${packagePath} has a "main" field set to "${main}", excluding the full filename and extension to the resolved file at "${urlPath.slice(packagePath.length)}", imported from ${basePath}.\n Automatic extension resolution of the "main" field is deprecated for ES modules.`, "DeprecationWarning", "DEP0151");
}
function tryStatSync(path$1) {
	try {
		return statSync(path$1);
	} catch {}
}
/**
* Legacy CommonJS main resolution:
* 1. let M = pkg_url + (json main field)
* 2. TRY(M, M.js, M.json, M.node)
* 3. TRY(M/index.js, M/index.json, M/index.node)
* 4. TRY(pkg_url/index.js, pkg_url/index.json, pkg_url/index.node)
* 5. NOT_FOUND
*/
function fileExists(url) {
	const stats = statSync(url, { throwIfNoEntry: false });
	const isFile = stats ? stats.isFile() : void 0;
	return isFile === null || isFile === void 0 ? false : isFile;
}
function legacyMainResolve(packageJsonUrl, packageConfig, base) {
	let guess;
	if (packageConfig.main !== void 0) {
		guess = new URL$1(packageConfig.main, packageJsonUrl);
		if (fileExists(guess)) return guess;
		const tries$1 = [
			`./${packageConfig.main}.js`,
			`./${packageConfig.main}.json`,
			`./${packageConfig.main}.node`,
			`./${packageConfig.main}/index.js`,
			`./${packageConfig.main}/index.json`,
			`./${packageConfig.main}/index.node`
		];
		let i$1 = -1;
		while (++i$1 < tries$1.length) {
			guess = new URL$1(tries$1[i$1], packageJsonUrl);
			if (fileExists(guess)) break;
			guess = void 0;
		}
		if (guess) {
			emitLegacyIndexDeprecation(guess, packageJsonUrl, base, packageConfig.main);
			return guess;
		}
	}
	const tries = [
		"./index.js",
		"./index.json",
		"./index.node"
	];
	let i = -1;
	while (++i < tries.length) {
		guess = new URL$1(tries[i], packageJsonUrl);
		if (fileExists(guess)) break;
		guess = void 0;
	}
	if (guess) {
		emitLegacyIndexDeprecation(guess, packageJsonUrl, base, packageConfig.main);
		return guess;
	}
	throw new ERR_MODULE_NOT_FOUND(fileURLToPath(new URL$1(".", packageJsonUrl)), fileURLToPath(base));
}
function finalizeResolution(resolved, base, preserveSymlinks) {
	if (encodedSeparatorRegEx.exec(resolved.pathname) !== null) throw new ERR_INVALID_MODULE_SPECIFIER(resolved.pathname, String.raw`must not include encoded "/" or "\" characters`, fileURLToPath(base));
	let filePath;
	try {
		filePath = fileURLToPath(resolved);
	} catch (error) {
		Object.defineProperty(error, "input", { value: String(resolved) });
		Object.defineProperty(error, "module", { value: String(base) });
		throw error;
	}
	const stats = tryStatSync(filePath.endsWith("/") ? filePath.slice(-1) : filePath);
	if (stats && stats.isDirectory()) {
		const error = new ERR_UNSUPPORTED_DIR_IMPORT(filePath, fileURLToPath(base));
		error.url = String(resolved);
		throw error;
	}
	if (!stats || !stats.isFile()) {
		const error = new ERR_MODULE_NOT_FOUND(filePath || resolved.pathname, base && fileURLToPath(base), true);
		error.url = String(resolved);
		throw error;
	}
	if (!preserveSymlinks) {
		const real = realpathSync(filePath);
		const { search, hash } = resolved;
		resolved = pathToFileURL(real + (filePath.endsWith(path.sep) ? "/" : ""));
		resolved.search = search;
		resolved.hash = hash;
	}
	return resolved;
}
function importNotDefined(specifier, packageJsonUrl, base) {
	return new ERR_PACKAGE_IMPORT_NOT_DEFINED(specifier, packageJsonUrl && fileURLToPath(new URL$1(".", packageJsonUrl)), fileURLToPath(base));
}
function exportsNotFound(subpath, packageJsonUrl, base) {
	return new ERR_PACKAGE_PATH_NOT_EXPORTED(fileURLToPath(new URL$1(".", packageJsonUrl)), subpath, base && fileURLToPath(base));
}
function throwInvalidSubpath(request, match, packageJsonUrl, internal, base) {
	throw new ERR_INVALID_MODULE_SPECIFIER(request, `request is not a valid match in pattern "${match}" for the "${internal ? "imports" : "exports"}" resolution of ${fileURLToPath(packageJsonUrl)}`, base && fileURLToPath(base));
}
function invalidPackageTarget(subpath, target, packageJsonUrl, internal, base) {
	target = typeof target === "object" && target !== null ? JSON.stringify(target, null, "") : `${target}`;
	return new ERR_INVALID_PACKAGE_TARGET(fileURLToPath(new URL$1(".", packageJsonUrl)), subpath, target, internal, base && fileURLToPath(base));
}
function resolvePackageTargetString(target, subpath, match, packageJsonUrl, base, pattern, internal, isPathMap, conditions) {
	if (subpath !== "" && !pattern && target.at(-1) !== "/") throw invalidPackageTarget(match, target, packageJsonUrl, internal, base);
	if (!target.startsWith("./")) {
		if (internal && !target.startsWith("../") && !target.startsWith("/")) {
			let isURL = false;
			try {
				new URL$1(target);
				isURL = true;
			} catch {}
			if (!isURL) return packageResolve(pattern ? RegExpPrototypeSymbolReplace.call(patternRegEx, target, () => subpath) : target + subpath, packageJsonUrl, conditions);
		}
		throw invalidPackageTarget(match, target, packageJsonUrl, internal, base);
	}
	if (invalidSegmentRegEx.exec(target.slice(2)) !== null) if (deprecatedInvalidSegmentRegEx.exec(target.slice(2)) === null) {
		if (!isPathMap) {
			const request = pattern ? match.replace("*", () => subpath) : match + subpath;
			emitInvalidSegmentDeprecation(pattern ? RegExpPrototypeSymbolReplace.call(patternRegEx, target, () => subpath) : target, request, match, packageJsonUrl, internal, base, true);
		}
	} else throw invalidPackageTarget(match, target, packageJsonUrl, internal, base);
	const resolved = new URL$1(target, packageJsonUrl);
	const resolvedPath = resolved.pathname;
	const packagePath = new URL$1(".", packageJsonUrl).pathname;
	if (!resolvedPath.startsWith(packagePath)) throw invalidPackageTarget(match, target, packageJsonUrl, internal, base);
	if (subpath === "") return resolved;
	if (invalidSegmentRegEx.exec(subpath) !== null) {
		const request = pattern ? match.replace("*", () => subpath) : match + subpath;
		if (deprecatedInvalidSegmentRegEx.exec(subpath) === null) {
			if (!isPathMap) emitInvalidSegmentDeprecation(pattern ? RegExpPrototypeSymbolReplace.call(patternRegEx, target, () => subpath) : target, request, match, packageJsonUrl, internal, base, false);
		} else throwInvalidSubpath(request, match, packageJsonUrl, internal, base);
	}
	if (pattern) return new URL$1(RegExpPrototypeSymbolReplace.call(patternRegEx, resolved.href, () => subpath));
	return new URL$1(subpath, resolved);
}
function isArrayIndex(key) {
	const keyNumber = Number(key);
	if (`${keyNumber}` !== key) return false;
	return keyNumber >= 0 && keyNumber < 4294967295;
}
function resolvePackageTarget(packageJsonUrl, target, subpath, packageSubpath, base, pattern, internal, isPathMap, conditions) {
	if (typeof target === "string") return resolvePackageTargetString(target, subpath, packageSubpath, packageJsonUrl, base, pattern, internal, isPathMap, conditions);
	if (Array.isArray(target)) {
		const targetList = target;
		if (targetList.length === 0) return null;
		let lastException;
		let i = -1;
		while (++i < targetList.length) {
			const targetItem = targetList[i];
			let resolveResult;
			try {
				resolveResult = resolvePackageTarget(packageJsonUrl, targetItem, subpath, packageSubpath, base, pattern, internal, isPathMap, conditions);
			} catch (error) {
				const exception = error;
				lastException = exception;
				if (exception.code === "ERR_INVALID_PACKAGE_TARGET") continue;
				throw error;
			}
			if (resolveResult === void 0) continue;
			if (resolveResult === null) {
				lastException = null;
				continue;
			}
			return resolveResult;
		}
		if (lastException === void 0 || lastException === null) return null;
		throw lastException;
	}
	if (typeof target === "object" && target !== null) {
		const keys = Object.getOwnPropertyNames(target);
		let i = -1;
		while (++i < keys.length) {
			const key = keys[i];
			if (isArrayIndex(key)) throw new ERR_INVALID_PACKAGE_CONFIG(fileURLToPath(packageJsonUrl), fileURLToPath(base), "\"exports\" cannot contain numeric property keys.");
		}
		i = -1;
		while (++i < keys.length) {
			const key = keys[i];
			if (key === "default" || conditions && conditions.has(key)) {
				const conditionalTarget = target[key];
				const resolveResult = resolvePackageTarget(packageJsonUrl, conditionalTarget, subpath, packageSubpath, base, pattern, internal, isPathMap, conditions);
				if (resolveResult === void 0) continue;
				return resolveResult;
			}
		}
		return null;
	}
	if (target === null) return null;
	throw invalidPackageTarget(packageSubpath, target, packageJsonUrl, internal, base);
}
function isConditionalExportsMainSugar(exports, packageJsonUrl, base) {
	if (typeof exports === "string" || Array.isArray(exports)) return true;
	if (typeof exports !== "object" || exports === null) return false;
	const keys = Object.getOwnPropertyNames(exports);
	let isConditionalSugar = false;
	let i = 0;
	let keyIndex = -1;
	while (++keyIndex < keys.length) {
		const key = keys[keyIndex];
		const currentIsConditionalSugar = key === "" || key[0] !== ".";
		if (i++ === 0) isConditionalSugar = currentIsConditionalSugar;
		else if (isConditionalSugar !== currentIsConditionalSugar) throw new ERR_INVALID_PACKAGE_CONFIG(fileURLToPath(packageJsonUrl), fileURLToPath(base), "\"exports\" cannot contain some keys starting with '.' and some not. The exports object must either be an object of package subpath keys or an object of main entry condition name keys only.");
	}
	return isConditionalSugar;
}
function emitTrailingSlashPatternDeprecation(match, pjsonUrl, base) {
	if (process$1.noDeprecation) return;
	const pjsonPath = fileURLToPath(pjsonUrl);
	if (emittedPackageWarnings.has(pjsonPath + "|" + match)) return;
	emittedPackageWarnings.add(pjsonPath + "|" + match);
	process$1.emitWarning(`Use of deprecated trailing slash pattern mapping "${match}" in the "exports" field module resolution of the package at ${pjsonPath}${base ? ` imported from ${fileURLToPath(base)}` : ""}. Mapping specifiers ending in "/" is no longer supported.`, "DeprecationWarning", "DEP0155");
}
function packageExportsResolve(packageJsonUrl, packageSubpath, packageConfig, base, conditions) {
	let exports = packageConfig.exports;
	if (isConditionalExportsMainSugar(exports, packageJsonUrl, base)) exports = { ".": exports };
	if (own.call(exports, packageSubpath) && !packageSubpath.includes("*") && !packageSubpath.endsWith("/")) {
		const target = exports[packageSubpath];
		const resolveResult = resolvePackageTarget(packageJsonUrl, target, "", packageSubpath, base, false, false, false, conditions);
		if (resolveResult === null || resolveResult === void 0) throw exportsNotFound(packageSubpath, packageJsonUrl, base);
		return resolveResult;
	}
	let bestMatch = "";
	let bestMatchSubpath = "";
	const keys = Object.getOwnPropertyNames(exports);
	let i = -1;
	while (++i < keys.length) {
		const key = keys[i];
		const patternIndex = key.indexOf("*");
		if (patternIndex !== -1 && packageSubpath.startsWith(key.slice(0, patternIndex))) {
			if (packageSubpath.endsWith("/")) emitTrailingSlashPatternDeprecation(packageSubpath, packageJsonUrl, base);
			const patternTrailer = key.slice(patternIndex + 1);
			if (packageSubpath.length >= key.length && packageSubpath.endsWith(patternTrailer) && patternKeyCompare(bestMatch, key) === 1 && key.lastIndexOf("*") === patternIndex) {
				bestMatch = key;
				bestMatchSubpath = packageSubpath.slice(patternIndex, packageSubpath.length - patternTrailer.length);
			}
		}
	}
	if (bestMatch) {
		const target = exports[bestMatch];
		const resolveResult = resolvePackageTarget(packageJsonUrl, target, bestMatchSubpath, bestMatch, base, true, false, packageSubpath.endsWith("/"), conditions);
		if (resolveResult === null || resolveResult === void 0) throw exportsNotFound(packageSubpath, packageJsonUrl, base);
		return resolveResult;
	}
	throw exportsNotFound(packageSubpath, packageJsonUrl, base);
}
function patternKeyCompare(a, b) {
	const aPatternIndex = a.indexOf("*");
	const bPatternIndex = b.indexOf("*");
	const baseLengthA = aPatternIndex === -1 ? a.length : aPatternIndex + 1;
	const baseLengthB = bPatternIndex === -1 ? b.length : bPatternIndex + 1;
	if (baseLengthA > baseLengthB) return -1;
	if (baseLengthB > baseLengthA) return 1;
	if (aPatternIndex === -1) return 1;
	if (bPatternIndex === -1) return -1;
	if (a.length > b.length) return -1;
	if (b.length > a.length) return 1;
	return 0;
}
function packageImportsResolve(name, base, conditions) {
	if (name === "#" || name.startsWith("#/") || name.endsWith("/")) throw new ERR_INVALID_MODULE_SPECIFIER(name, "is not a valid internal imports specifier name", fileURLToPath(base));
	let packageJsonUrl;
	const packageConfig = getPackageScopeConfig(base);
	if (packageConfig.exists) {
		packageJsonUrl = pathToFileURL(packageConfig.pjsonPath);
		const imports = packageConfig.imports;
		if (imports) if (own.call(imports, name) && !name.includes("*")) {
			const resolveResult = resolvePackageTarget(packageJsonUrl, imports[name], "", name, base, false, true, false, conditions);
			if (resolveResult !== null && resolveResult !== void 0) return resolveResult;
		} else {
			let bestMatch = "";
			let bestMatchSubpath = "";
			const keys = Object.getOwnPropertyNames(imports);
			let i = -1;
			while (++i < keys.length) {
				const key = keys[i];
				const patternIndex = key.indexOf("*");
				if (patternIndex !== -1 && name.startsWith(key.slice(0, -1))) {
					const patternTrailer = key.slice(patternIndex + 1);
					if (name.length >= key.length && name.endsWith(patternTrailer) && patternKeyCompare(bestMatch, key) === 1 && key.lastIndexOf("*") === patternIndex) {
						bestMatch = key;
						bestMatchSubpath = name.slice(patternIndex, name.length - patternTrailer.length);
					}
				}
			}
			if (bestMatch) {
				const target = imports[bestMatch];
				const resolveResult = resolvePackageTarget(packageJsonUrl, target, bestMatchSubpath, bestMatch, base, true, true, false, conditions);
				if (resolveResult !== null && resolveResult !== void 0) return resolveResult;
			}
		}
	}
	throw importNotDefined(name, packageJsonUrl, base);
}
/**
* @param {string} specifier
* @param {URL} base
*/
function parsePackageName(specifier, base) {
	let separatorIndex = specifier.indexOf("/");
	let validPackageName = true;
	let isScoped = false;
	if (specifier[0] === "@") {
		isScoped = true;
		if (separatorIndex === -1 || specifier.length === 0) validPackageName = false;
		else separatorIndex = specifier.indexOf("/", separatorIndex + 1);
	}
	const packageName = separatorIndex === -1 ? specifier : specifier.slice(0, separatorIndex);
	if (invalidPackageNameRegEx.exec(packageName) !== null) validPackageName = false;
	if (!validPackageName) throw new ERR_INVALID_MODULE_SPECIFIER(specifier, "is not a valid package name", fileURLToPath(base));
	return {
		packageName,
		packageSubpath: "." + (separatorIndex === -1 ? "" : specifier.slice(separatorIndex)),
		isScoped
	};
}
function packageResolve(specifier, base, conditions) {
	if (nodeBuiltins.includes(specifier)) return new URL$1("node:" + specifier);
	const { packageName, packageSubpath, isScoped } = parsePackageName(specifier, base);
	const packageConfig = getPackageScopeConfig(base);
	if (packageConfig.exists && packageConfig.name === packageName && packageConfig.exports !== void 0 && packageConfig.exports !== null) return packageExportsResolve(pathToFileURL(packageConfig.pjsonPath), packageSubpath, packageConfig, base, conditions);
	let packageJsonUrl = new URL$1("./node_modules/" + packageName + "/package.json", base);
	let packageJsonPath = fileURLToPath(packageJsonUrl);
	let lastPath;
	do {
		const stat = tryStatSync(packageJsonPath.slice(0, -13));
		if (!stat || !stat.isDirectory()) {
			lastPath = packageJsonPath;
			packageJsonUrl = new URL$1((isScoped ? "../../../../node_modules/" : "../../../node_modules/") + packageName + "/package.json", packageJsonUrl);
			packageJsonPath = fileURLToPath(packageJsonUrl);
			continue;
		}
		const packageConfig$1 = read(packageJsonPath, {
			base,
			specifier
		});
		if (packageConfig$1.exports !== void 0 && packageConfig$1.exports !== null) return packageExportsResolve(packageJsonUrl, packageSubpath, packageConfig$1, base, conditions);
		if (packageSubpath === ".") return legacyMainResolve(packageJsonUrl, packageConfig$1, base);
		return new URL$1(packageSubpath, packageJsonUrl);
	} while (packageJsonPath.length !== lastPath.length);
	throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), false);
}
function isRelativeSpecifier(specifier) {
	if (specifier[0] === ".") {
		if (specifier.length === 1 || specifier[1] === "/") return true;
		if (specifier[1] === "." && (specifier.length === 2 || specifier[2] === "/")) return true;
	}
	return false;
}
function shouldBeTreatedAsRelativeOrAbsolutePath(specifier) {
	if (specifier === "") return false;
	if (specifier[0] === "/") return true;
	return isRelativeSpecifier(specifier);
}
/**
* The “Resolver Algorithm Specification” as detailed in the Node docs (which is
* sync and slightly lower-level than `resolve`).
*
* @param {string} specifier
*   `/example.js`, `./example.js`, `../example.js`, `some-package`, `fs`, etc.
* @param {URL} base
*   Full URL (to a file) that `specifier` is resolved relative from.
* @param {Set<string>} [conditions]
*   Conditions.
* @param {boolean} [preserveSymlinks]
*   Keep symlinks instead of resolving them.
* @returns {URL}
*   A URL object to the found thing.
*/
function moduleResolve(specifier, base, conditions, preserveSymlinks) {
	const protocol = base.protocol;
	const isData = protocol === "data:";
	let resolved;
	if (shouldBeTreatedAsRelativeOrAbsolutePath(specifier)) try {
		resolved = new URL$1(specifier, base);
	} catch (error_) {
		const error = new ERR_UNSUPPORTED_RESOLVE_REQUEST(specifier, base);
		error.cause = error_;
		throw error;
	}
	else if (protocol === "file:" && specifier[0] === "#") resolved = packageImportsResolve(specifier, base, conditions);
	else try {
		resolved = new URL$1(specifier);
	} catch (error_) {
		if (isData && !nodeBuiltins.includes(specifier)) {
			const error = new ERR_UNSUPPORTED_RESOLVE_REQUEST(specifier, base);
			error.cause = error_;
			throw error;
		}
		resolved = packageResolve(specifier, base, conditions);
	}
	assert.ok(resolved !== void 0, "expected to be defined");
	if (resolved.protocol !== "file:") return resolved;
	return finalizeResolution(resolved, base, preserveSymlinks);
}
const DEFAULT_CONDITIONS_SET = /* @__PURE__ */ new Set(["node", "import"]);
const isWindows$1 = process.platform === "win32";
const globalCache = globalThis["__EXSOLVE_CACHE__"] ||= /* @__PURE__ */ new Map();
/**
* Synchronously resolves a module url based on the options provided.
*
* @param {string} input - The identifier or path of the module to resolve.
* @param {ResolveOptions} [options] - Options to resolve the module. See {@link ResolveOptions}.
* @returns {string} The resolved URL as a string.
*/
function resolveModuleURL(input, options) {
	const parsedInput = _parseInput(input);
	if ("external" in parsedInput) return parsedInput.external;
	const specifier = parsedInput.specifier;
	let url = parsedInput.url;
	let absolutePath = parsedInput.absolutePath;
	let cacheKey;
	let cacheObj;
	if (options?.cache !== false) {
		cacheKey = _cacheKey(absolutePath || specifier, options);
		cacheObj = options?.cache && typeof options?.cache === "object" ? options.cache : globalCache;
	}
	if (cacheObj) {
		const cached = cacheObj.get(cacheKey);
		if (typeof cached === "string") return cached;
		if (cached instanceof Error) {
			if (options?.try) return;
			throw cached;
		}
	}
	if (absolutePath) try {
		const stat = lstatSync(absolutePath);
		if (stat.isSymbolicLink()) {
			absolutePath = realpathSync(absolutePath);
			url = pathToFileURL(absolutePath);
		}
		if (stat.isFile()) {
			if (cacheObj) cacheObj.set(cacheKey, url.href);
			return url.href;
		}
	} catch (error) {
		if (error?.code !== "ENOENT") {
			if (cacheObj) cacheObj.set(cacheKey, error);
			throw error;
		}
	}
	const conditionsSet = options?.conditions ? new Set(options.conditions) : DEFAULT_CONDITIONS_SET;
	const target = specifier || url.href;
	const bases = _normalizeBases(options?.from);
	const suffixes = options?.suffixes || [""];
	const extensions = options?.extensions ? ["", ...options.extensions] : [""];
	let resolved;
	for (const base of bases) {
		for (const suffix of suffixes) {
			let name = _join(target, suffix);
			if (name === ".") name += "/.";
			for (const extension of extensions) {
				resolved = _tryModuleResolve(name + extension, base, conditionsSet);
				if (resolved) break;
			}
			if (resolved) break;
		}
		if (resolved) break;
	}
	if (!resolved) {
		const error = /* @__PURE__ */ new Error(`Cannot resolve module "${input}" (from: ${bases.map((u) => _fmtPath(u)).join(", ")})`);
		error.code = "ERR_MODULE_NOT_FOUND";
		if (cacheObj) cacheObj.set(cacheKey, error);
		if (options?.try) return;
		throw error;
	}
	if (cacheObj) cacheObj.set(cacheKey, resolved.href);
	return resolved.href;
}
/**
* Synchronously resolves a module then converts it to a file path
*
* (throws error if reolved path is not file:// scheme)
*
* @param {string} id - The identifier or path of the module to resolve.
* @param {ResolveOptions} [options] - Options to resolve the module. See {@link ResolveOptions}.
* @returns {string} The resolved URL as a string.
*/
function resolveModulePath(id, options) {
	const resolved = resolveModuleURL(id, options);
	if (!resolved) return;
	if (!resolved.startsWith("file://") && options?.try) return;
	const absolutePath = fileURLToPath(resolved);
	return isWindows$1 ? _normalizeWinPath(absolutePath) : absolutePath;
}
function _tryModuleResolve(specifier, base, conditions) {
	try {
		return moduleResolve(specifier, base, conditions);
	} catch {}
}
function _normalizeBases(inputs) {
	const urls = (Array.isArray(inputs) ? inputs : [inputs]).flatMap((input) => _normalizeBase(input));
	if (urls.length === 0) return [pathToFileURL("./")];
	return urls;
}
function _normalizeBase(input) {
	if (!input) return [];
	if (_isURL(input)) return [input];
	if (typeof input !== "string") return [];
	if (/^(?:node|data|http|https|file):/.test(input)) return new URL(input);
	try {
		if (input.endsWith("/") || statSync(input).isDirectory()) return pathToFileURL(input + "/");
		return pathToFileURL(input);
	} catch {
		return [pathToFileURL(input + "/"), pathToFileURL(input)];
	}
}
function _fmtPath(input) {
	try {
		return fileURLToPath(input);
	} catch {
		return input;
	}
}
function _cacheKey(id, opts) {
	return JSON.stringify([
		id,
		(opts?.conditions || ["node", "import"]).sort(),
		opts?.extensions,
		opts?.from,
		opts?.suffixes
	]);
}
function _join(a, b) {
	if (!a || !b || b === "/") return a;
	return (a.endsWith("/") ? a : a + "/") + (b.startsWith("/") ? b.slice(1) : b);
}
function _normalizeWinPath(path$1) {
	return path$1.replace(/\\/g, "/").replace(/^[a-z]:\//, (r) => r.toUpperCase());
}
function _isURL(input) {
	return input instanceof URL || input?.constructor?.name === "URL";
}
function _parseInput(input) {
	if (typeof input === "string") {
		if (input.startsWith("file:")) {
			const url = new URL(input);
			return {
				url,
				absolutePath: fileURLToPath(url)
			};
		}
		if (isAbsolute(input)) return {
			url: pathToFileURL(input),
			absolutePath: input
		};
		if (/^(?:node|data|http|https):/.test(input)) return { external: input };
		if (nodeBuiltins.includes(input) && !input.includes(":")) return { external: `node:${input}` };
		return { specifier: input };
	}
	if (_isURL(input)) {
		if (input.protocol === "file:") return {
			url: input,
			absolutePath: fileURLToPath(input)
		};
		return { external: input.href };
	}
	throw new TypeError("id must be a `string` or `URL`");
}
//#endregion
//#region node_modules/.pnpm/pkg-types@2.3.0/node_modules/pkg-types/dist/index.mjs
const defaultFindOptions = {
	startingFrom: ".",
	rootPattern: /^node_modules$/,
	reverse: false,
	test: (filePath) => {
		try {
			if (statSync(filePath).isFile()) return true;
		} catch {}
	}
};
async function findFile(filename, _options = {}) {
	const filenames = Array.isArray(filename) ? filename : [filename];
	const options = {
		...defaultFindOptions,
		..._options
	};
	const basePath = resolve$1(options.startingFrom);
	const leadingSlash = basePath[0] === "/";
	const segments = basePath.split("/").filter(Boolean);
	if (filenames.includes(segments.at(-1)) && await options.test(basePath)) return basePath;
	if (leadingSlash) segments[0] = "/" + segments[0];
	let root = segments.findIndex((r) => r.match(options.rootPattern));
	if (root === -1) root = 0;
	if (options.reverse) for (let index = root + 1; index <= segments.length; index++) for (const filename2 of filenames) {
		const filePath = join$1(...segments.slice(0, index), filename2);
		if (await options.test(filePath)) return filePath;
	}
	else for (let index = segments.length; index > root; index--) for (const filename2 of filenames) {
		const filePath = join$1(...segments.slice(0, index), filename2);
		if (await options.test(filePath)) return filePath;
	}
	throw new Error(`Cannot find matching ${filename} in ${options.startingFrom} or parent directories`);
}
function findNearestFile(filename, options = {}) {
	return findFile(filename, options);
}
function _resolvePath(id, opts = {}) {
	if (id instanceof URL || id.startsWith("file://")) return normalize(fileURLToPath(id));
	if (isAbsolute$1(id)) return normalize(id);
	return resolveModulePath(id, {
		...opts,
		from: opts.from || opts.parent || opts.url
	});
}
const FileCache = /* @__PURE__ */ new Map();
async function readPackageJSON(id, options = {}) {
	const resolvedPath = await resolvePackageJSON(id, options);
	const cache = options.cache && typeof options.cache !== "boolean" ? options.cache : FileCache;
	if (options.cache && cache.has(resolvedPath)) return cache.get(resolvedPath);
	const blob = await promises.readFile(resolvedPath, "utf8");
	let parsed;
	try {
		parsed = n(blob);
	} catch {
		parsed = r(blob);
	}
	cache.set(resolvedPath, parsed);
	return parsed;
}
async function resolvePackageJSON(id = process.cwd(), options = {}) {
	return findNearestFile("package.json", {
		...options,
		startingFrom: _resolvePath(id, options)
	});
}
//#endregion
//#region node_modules/.pnpm/changelogen@0.6.2/node_modules/changelogen/dist/shared/changelogen.D-9f3HTX.mjs
function execCommand(cmd, cwd) {
	return execSync(cmd, {
		encoding: "utf8",
		cwd
	}).trim();
}
async function getLastGitTag(cwd) {
	try {
		return execCommand("git describe --tags --abbrev=0", cwd)?.split("\n").at(-1);
	} catch {}
}
function getCurrentGitBranch(cwd) {
	return execCommand("git rev-parse --abbrev-ref HEAD", cwd);
}
function getCurrentGitTag(cwd) {
	return execCommand("git tag --points-at HEAD", cwd);
}
function getCurrentGitRef(cwd) {
	return getCurrentGitTag(cwd) || getCurrentGitBranch(cwd);
}
function getGitRemoteURL(cwd, remote = "origin") {
	return execCommand(`git --work-tree="${cwd}" remote get-url "${remote}"`, cwd);
}
async function getGitDiff(from, to = "HEAD", cwd) {
	return execCommand(`git --no-pager log "${from ? `${from}...` : ""}${to}" --pretty="----%n%s|%h|%an|%ae%n%b" --name-status`, cwd).split("----\n").splice(1).map((line) => {
		const [firstLine, ..._body] = line.split("\n");
		const [message, shortHash, authorName, authorEmail] = firstLine.split("|");
		return {
			message,
			shortHash,
			author: {
				name: authorName,
				email: authorEmail
			},
			body: _body.join("\n")
		};
	});
}
function parseCommits(commits, config) {
	return commits.map((commit) => parseGitCommit(commit, config)).filter(Boolean);
}
const ConventionalCommitRegex = /(?<emoji>:.+:|(\uD83C[\uDF00-\uDFFF])|(\uD83D[\uDC00-\uDE4F\uDE80-\uDEFF])|[\u2600-\u2B55])?( *)?(?<type>[a-z]+)(\((?<scope>.+)\))?(?<breaking>!)?: (?<description>.+)/i;
const CoAuthoredByRegex = /co-authored-by:\s*(?<name>.+)(<(?<email>.+)>)/gim;
const PullRequestRE = /\([ a-z]*(#\d+)\s*\)/gm;
const IssueRE = /(#\d+)/gm;
function parseGitCommit(commit, config) {
	const match = commit.message.match(ConventionalCommitRegex);
	if (!match) return null;
	const type = match.groups.type;
	const hasBreakingBody = /breaking change:/i.test(commit.body);
	let scope = match.groups.scope || "";
	scope = config.scopeMap[scope] || scope;
	const isBreaking = Boolean(match.groups.breaking || hasBreakingBody);
	let description = match.groups.description;
	const references = [];
	for (const m of description.matchAll(PullRequestRE)) references.push({
		type: "pull-request",
		value: m[1]
	});
	for (const m of description.matchAll(IssueRE)) if (!references.some((i) => i.value === m[1])) references.push({
		type: "issue",
		value: m[1]
	});
	references.push({
		value: commit.shortHash,
		type: "hash"
	});
	description = description.replace(PullRequestRE, "").trim();
	const authors = [commit.author];
	for (const match2 of commit.body.matchAll(CoAuthoredByRegex)) authors.push({
		name: (match2.groups.name || "").trim(),
		email: (match2.groups.email || "").trim()
	});
	return {
		...commit,
		authors,
		description,
		type,
		scope,
		references,
		isBreaking
	};
}
const providerToRefSpec = {
	github: {
		"pull-request": "pull",
		hash: "commit",
		issue: "issues"
	},
	gitlab: {
		"pull-request": "merge_requests",
		hash: "commit",
		issue: "issues"
	},
	bitbucket: {
		"pull-request": "pull-requests",
		hash: "commit",
		issue: "issues"
	}
};
const providerToDomain = {
	github: "github.com",
	gitlab: "gitlab.com",
	bitbucket: "bitbucket.org"
};
const domainToProvider = {
	"github.com": "github",
	"gitlab.com": "gitlab",
	"bitbucket.org": "bitbucket"
};
const providerURLRegex = /^(?:(?<user>[\w-]+)@)?(?:(?<provider>[^/:]+):)?(?<repo>[\w-]+\/(?:\w|\.(?!git$)|-)+)(?:\.git)?$/;
function baseUrl(config) {
	return `https://${config.domain}/${config.repo}`;
}
function formatReference(ref, repo) {
	if (!repo || !(repo.provider in providerToRefSpec)) return ref.value;
	const refSpec = providerToRefSpec[repo.provider];
	return `[${ref.value}](${baseUrl(repo)}/${refSpec[ref.type]}/${ref.value.replace(/^#/, "")})`;
}
function formatCompareChanges(v, config) {
	const part = config.repo.provider === "bitbucket" ? "branches/compare" : "compare";
	const changes = config.repo.provider === "bitbucket" ? `${v || config.to}%0D${config.from}` : `${config.from}...${v || config.to}`;
	return `[compare changes](${baseUrl(config.repo)}/${part}/${changes})`;
}
async function resolveRepoConfig(cwd) {
	const pkg = await readPackageJSON(cwd).catch(() => {});
	if (pkg && pkg.repository) return getRepoConfig(typeof pkg.repository === "string" ? pkg.repository : pkg.repository.url);
	try {
		const gitRemote = getGitRemoteURL(cwd);
		if (gitRemote) return getRepoConfig(gitRemote);
	} catch {}
}
function getRepoConfig(repoUrl = "") {
	let provider;
	let repo;
	let domain;
	let url;
	try {
		url = new URL(repoUrl);
	} catch {}
	const m = repoUrl.match(providerURLRegex)?.groups ?? {};
	if (m.repo && m.provider) {
		repo = m.repo;
		provider = m.provider in domainToProvider ? domainToProvider[m.provider] : m.provider;
		domain = provider in providerToDomain ? providerToDomain[provider] : provider;
	} else if (url) {
		domain = url.hostname;
		repo = url.pathname.split("/").slice(1).join("/").replace(/\.git$/, "");
		provider = domainToProvider[domain];
	} else if (m.repo) {
		repo = m.repo;
		provider = "github";
		domain = providerToDomain[provider];
	}
	return {
		provider,
		repo,
		domain
	};
}
async function generateMarkDown(commits, config) {
	const typeGroups = groupBy(commits, "type");
	const markdown = [];
	const breakingChanges = [];
	const v = config.newVersion && config.templates.tagBody.replaceAll("{{newVersion}}", config.newVersion);
	markdown.push("", "## " + (v || `${config.from || ""}...${config.to}`), "");
	if (config.repo && config.from) markdown.push(formatCompareChanges(v, config));
	for (const type in config.types) {
		const group = typeGroups[type];
		if (!group || group.length === 0) continue;
		markdown.push("", "### " + config.types[type].title, "");
		for (const commit of group.reverse()) {
			const line = formatCommit(commit, config);
			markdown.push(line);
			if (commit.isBreaking) breakingChanges.push(line);
		}
	}
	if (breakingChanges.length > 0) markdown.push("", "#### ⚠️ Breaking Changes", "", ...breakingChanges);
	const _authors = /* @__PURE__ */ new Map();
	for (const commit of commits) {
		if (!commit.author) continue;
		const name = formatName(commit.author.name);
		if (!name || name.includes("[bot]")) continue;
		if (config.excludeAuthors && config.excludeAuthors.some((v2) => name.includes(v2) || commit.author.email?.includes(v2))) continue;
		if (_authors.has(name)) _authors.get(name).email.add(commit.author.email);
		else _authors.set(name, { email: /* @__PURE__ */ new Set([commit.author.email]) });
	}
	await Promise.all([..._authors.keys()].map(async (authorName) => {
		const meta = _authors.get(authorName);
		for (const email of meta.email) {
			const { user } = await l(`https://ungh.cc/users/find/${email}`).then((r) => r.json()).catch(() => ({ user: null }));
			if (user) {
				meta.github = user.username;
				break;
			}
		}
	}));
	const authors = [..._authors.entries()].map((e) => ({
		name: e[0],
		...e[1]
	}));
	if (authors.length > 0 && !config.noAuthors) markdown.push("", "### ❤️ Contributors", "", ...authors.map((i) => {
		const _email = [...i.email].find((e) => !e.includes("noreply.github.com"));
		const email = config.hideAuthorEmail !== true && _email ? ` <${_email}>` : "";
		const github = i.github ? ` ([@${i.github}](https://github.com/${i.github}))` : "";
		return `- ${i.name}${github || email || ""}`;
	}));
	return convert(markdown.join("\n").trim(), true);
}
function formatCommit(commit, config) {
	return "- " + (commit.scope ? `**${commit.scope.trim()}:** ` : "") + (commit.isBreaking ? "⚠️  " : "") + upperFirst(commit.description) + formatReferences(commit.references, config);
}
function formatReferences(references, config) {
	const pr = references.filter((ref) => ref.type === "pull-request");
	const issue = references.filter((ref) => ref.type === "issue");
	if (pr.length > 0 || issue.length > 0) return " (" + [...pr, ...issue].map((ref) => formatReference(ref, config.repo)).join(", ") + ")";
	if (references.length > 0) return " (" + formatReference(references[0], config.repo) + ")";
	return "";
}
function formatName(name = "") {
	return name.split(" ").map((p) => upperFirst(p.trim())).join(" ");
}
function groupBy(items, key) {
	const groups = {};
	for (const item of items) {
		groups[item[key]] = groups[item[key]] || [];
		groups[item[key]].push(item);
	}
	return groups;
}
const defaultOutput = "CHANGELOG.md";
const getDefaultConfig = () => ({
	types: {
		feat: {
			title: "🚀 Enhancements",
			semver: "minor"
		},
		perf: {
			title: "🔥 Performance",
			semver: "patch"
		},
		fix: {
			title: "🩹 Fixes",
			semver: "patch"
		},
		refactor: {
			title: "💅 Refactors",
			semver: "patch"
		},
		docs: {
			title: "📖 Documentation",
			semver: "patch"
		},
		build: {
			title: "📦 Build",
			semver: "patch"
		},
		types: {
			title: "🌊 Types",
			semver: "patch"
		},
		chore: { title: "🏡 Chore" },
		examples: { title: "🏀 Examples" },
		test: { title: "✅ Tests" },
		style: { title: "🎨 Styles" },
		ci: { title: "🤖 CI" }
	},
	cwd: null,
	from: "",
	to: "",
	output: defaultOutput,
	scopeMap: {},
	tokens: { github: process.env.CHANGELOGEN_TOKENS_GITHUB || process.env.GITHUB_TOKEN || process.env.GH_TOKEN },
	publish: {
		private: false,
		tag: "latest",
		args: []
	},
	templates: {
		commitMessage: "chore(release): v{{newVersion}}",
		tagMessage: "v{{newVersion}}",
		tagBody: "v{{newVersion}}"
	},
	excludeAuthors: [],
	noAuthors: false
});
async function loadChangelogConfig(cwd, overrides) {
	await setupDotenv({ cwd });
	const { config } = await loadConfig({
		cwd,
		name: "changelog",
		packageJson: true,
		defaults: getDefaultConfig(),
		overrides: {
			cwd,
			...overrides
		}
	});
	return await resolveChangelogConfig(config, cwd);
}
async function resolveChangelogConfig(config, cwd) {
	if (!config.from) config.from = await getLastGitTag(cwd);
	if (!config.to) config.to = await getCurrentGitRef(cwd);
	if (config.output) config.output = config.output === true ? defaultOutput : resolve(cwd, config.output);
	else config.output = false;
	if (!config.repo) config.repo = await resolveRepoConfig(cwd);
	if (typeof config.repo === "string") config.repo = getRepoConfig(config.repo);
	return config;
}
//#endregion
//#region src/core/releaser/changelog.ts
async function getConventionalChangelog(commits) {
	if (!commits[0].type) return commits.map((c) => c.message).join("\n");
	return (await generateMarkDown(commits.map((c) => {
		if (c.type === "remove") c.isBreaking = true;
		return c;
	}), await loadChangelogConfig(process$1.cwd(), {
		types: {
			add: {
				title: "🚀 Enhancements",
				semver: "minor"
			},
			change: {
				title: "🩹 Fixes",
				semver: "patch"
			},
			remove: {
				title: "🩹 Fixes",
				semver: "minor"
			}
		},
		hideAuthorEmail: true
	}))).split("\n").slice(3).join("\n");
}
async function getGitCommits(currentTag, commitMessageTemplate) {
	/**
	* Get all git tags
	*
	* @example
	*
	* ```bash
	* $ git tag -l --sort=v:refname
	* v2.0.9
	* v2.0.10
	* v2.0.11
	* v2.0.12
	* v2.0.13
	* v2.0.13-beta.1
	* v2.0.13-beta.2
	* v2.0.13-beta.3
	* v2.0.14
	* ```
	*/
	const tags = execSync("git tag -l --sort=v:refname").toString().trim().split("\n");
	const currentTagIndex = tags.indexOf(currentTag);
	if (currentTagIndex === -1) throw new Error(`Tag "${currentTag}" not found.`);
	let previousTagIndex = currentTagIndex - 1;
	let previousTag;
	if (currentTagIndex === 0) previousTag = void 0;
	else if (currentTag.includes("-")) previousTag = tags[previousTagIndex];
	else {
		while (previousTagIndex >= 0 && tags[previousTagIndex].includes("-")) previousTagIndex--;
		if (previousTagIndex < 0) previousTag = void 0;
		else previousTag = tags[previousTagIndex];
	}
	const filterRegex = new RegExp(escapeRegExp(commitMessageTemplate));
	return parseCommits((await getGitDiff(previousTag, currentTag)).filter((c) => !filterRegex.test(c.message)), { scopeMap: {} });
}
//#endregion
//#region src/utils/mime.ts
const mimeTypes = {
	json: "application/json",
	xpi: "application/x-xpinstall"
};
function getMimeTypeByFileName(filename) {
	const ext = extname(filename);
	for (const type in mimeTypes) if (ext === `.${type}`) return mimeTypes[type];
}
//#endregion
//#region src/core/releaser/github.ts
var GitHub = class extends ReleaseBase {
	client;
	constructor(ctx) {
		super(ctx);
		this.client = this.getClient();
	}
	async run() {
		this.checkFiles();
		this.logger.info("Uploading XPI to GitHub...");
		await this.uploadXPI();
		this.logger.info("Refreshing update manifest...");
		await this.refreshUpdateManifest();
		return this.ctx;
	}
	/**
	* Create new release and upload XPI to asset
	*/
	async uploadXPI() {
		const { version, dist, xpiName } = this.ctx;
		const release = await this.createRelease({
			...this.remote,
			tag_name: this.ctx.release.bumpp.tag.toString().replaceAll("%s", version),
			name: `Release v${version}`,
			body: await this.getChangelog(),
			prerelease: version.includes("-"),
			make_latest: "true"
		});
		if (!release) throw new Error("Create release failed!");
		this.logger.debug("Uploading xpi asset...");
		await this.uploadAsset(release.id, join(dist, `${xpiName}.xpi`));
	}
	async getReleaseByTag(tag) {
		return await this.client.rest.repos.getReleaseByTag({
			...this.remote,
			tag
		}).catch((e) => {
			this.logger.debug(`Release with tag ${tag} not found. ${e}`);
		}).then((res) => {
			if (res && res.status === 200) {
				this.logger.debug(`Found release with tag "${tag}", id=${res.data.id}.`);
				return res.data;
			}
		});
	}
	async createRelease(options) {
		this.logger.debug("Creating release...");
		this.logger.debug(options);
		return await this.client.rest.repos.createRelease(options).catch((e) => {
			this.logger.error(e);
			throw new Error("Create release failed.");
		}).then((res) => {
			if (res.status === 201) {
				this.logger.debug(`Create release "${res.data.tag_name}" success, id: ${res.data.id}.`);
				return res.data;
			}
		});
	}
	async uploadAsset(releaseID, asset) {
		this.logger.debug(`Uploading ${asset} to release ${releaseID}`);
		return await this.client.rest.repos.uploadReleaseAsset({
			...this.remote,
			release_id: releaseID,
			data: await readFile(asset),
			headers: {
				"content-type": getMimeTypeByFileName(asset) || "application/octet-stream",
				"content-length": (await stat(asset)).size
			},
			name: basename(asset)
		}).then((res) => {
			this.logger.debug(`Upload "${res.data.name}" success, assetId: ${res.data.id}`);
			return res.data;
		});
	}
	async refreshUpdateManifest() {
		const updater = this.ctx.release.github.updater;
		if (!updater) {
			this.logger.debug(`Skip refresh update.json because release.github.updater = false`);
			return;
		}
		const { dist, version } = this.ctx;
		const assets = (await glob(`${dist}/update*.json`)).map((p) => basename(p));
		const release = await this.getReleaseByTag(updater) ?? await this.createRelease({
			...this.remote,
			tag_name: updater,
			prerelease: true,
			make_latest: "false"
		});
		if (!release) throw new Error("Get or create 'release' failed.");
		const existAssets = await this.client.rest.repos.listReleaseAssets({
			...this.remote,
			release_id: release.id
		}).then((res) => {
			return res.data.filter((asset) => assets.includes(asset.name));
		});
		if (existAssets) {
			for (const existAsset of existAssets) if (assets.includes(existAsset.name)) {
				this.logger.debug(`Delete existed asset ${existAsset.name} in release ${updater}`);
				await this.client.rest.repos.deleteReleaseAsset({
					...this.remote,
					asset_id: existAsset.id
				});
			}
		}
		for (const asset of assets) await this.uploadAsset(release.id, join(dist, asset));
		await this.client.rest.repos.updateRelease({
			...this.remote,
			release_id: release.id,
			name: "Release Manifest",
			body: `This release is used to host \`update.json\`, please do not delete or modify it! \n Updated in UTC ${(/* @__PURE__ */ new Date()).toISOString()} for version ${version}`,
			prerelease: true,
			make_latest: "false"
		});
	}
	async getChangelog() {
		const { release } = this.ctx;
		const { github } = release;
		const { releaseNote } = github;
		return releaseNote(this.ctx);
	}
	getClient() {
		if (!process$1.env.GITHUB_TOKEN) throw new Error("No GITHUB_TOKEN.");
		return new Octokit({
			auth: process$1.env.GITHUB_TOKEN,
			userAgent: "zotero-plugin-scaffold"
		});
	}
	get remote() {
		const [owner, repo] = this.ctx.release.github.repository.split("/");
		return {
			owner,
			repo
		};
	}
};
//#endregion
//#region src/core/releaser/index.ts
var Release = class extends Base {
	constructor(ctx) {
		super(ctx);
	}
	/**
	* Runs release
	*
	* if is not CI，bump version, git add (package.json), git commit, git tag, git push;
	* if is CI, do not bump version, do not run git, create release (tag is `v${version}`) and upload xpi,
	*    then, create or update release (tag is "release"), update `update.json`.
	*/
	async run() {
		const t = /* @__PURE__ */ new Date();
		const { release, version } = this.ctx;
		if (release.bumpp.release === "prompt" && isCI) {
			this.logger.warn("Config `release.bumpp.release == 'prompt'` will do nothing because in CI enviroment.");
			this.ctx.release.bumpp.release = version;
		}
		if (release.bumpp.confirm && isCI) {
			this.logger.warn("Config `release.bumpp.confirm` will do nothing because in CI enviroment.");
			this.ctx.release.bumpp.confirm = false;
		}
		const isBumpNeeded = this.ctx.release.bumpp.release !== version;
		const isGitHubEnabled = this.isEnabled(release.github.enable);
		if (isGitHubEnabled && !release.bumpp.execute) {
			this.logger.warn(`The current release needs to run the build after bumping the version number, please configure the build script in 'config.release.bumpp.execute'${isBumpNeeded ? "" : " or run build before run release"}.`);
			this.ctx.release.bumpp.execute ||= "npm run build";
		}
		this.logger.debug(`Release config: ", ${this.ctx.release}`);
		await this.ctx.hooks.callHook("release:init", this.ctx);
		await new Bump(this.ctx).run();
		await this.ctx.hooks.callHook("release:push", this.ctx);
		this.ctx.release.changelog = await this.getChangelog();
		if (isGitHubEnabled) await new GitHub(this.ctx).run();
		await this.ctx.hooks.callHook("release:done", this.ctx);
		this.logger.success(`Done in ${(Date.now() - t.getTime()) / 1e3} s.`);
	}
	async getChangelog() {
		const { commit, tag } = this.ctx.release.bumpp;
		let changelog;
		const rawCommit = await getGitCommits(tag, commit);
		if (rawCommit.length === 0) return "_No significant changes._";
		const changelogConfig = this.ctx.release.changelog;
		if (typeof changelogConfig == "function") changelog = changelogConfig(this.ctx, rawCommit);
		else if (!!changelogConfig && typeof changelogConfig == "string") changelog = execSync(changelogConfig).toString().trim();
		else changelog = await getConventionalChangelog(rawCommit);
		this.logger.debug(`Got changelog:\n${changelog}\n`);
		return changelog;
	}
	isEnabled(enable) {
		if (enable === "always") return true;
		if (enable === "ci" && isCI) return true;
		if (enable === "local" && !isCI) return true;
		return false;
	}
	exit() {}
	get resolvedCommitMessage() {
		return this.ctx.release.bumpp.commit.replace("%s", this.ctx.version);
	}
};
//#endregion
//#region src/utils/watcher.ts
function watch(source, ignore, event) {
	const watcher = chokidar.watch(source, {
		ignored: [
			/(^|[/\\])\../,
			/[\\/]\.git[\\/]/,
			/[\\/]node_modules[\\/]/,
			...toArray(ignore)
		],
		ignoreInitial: true,
		ignorePermissionErrors: true,
		persistent: true
	});
	const onChangeDebounced = safeDebounce(event.onChange);
	const onAddDebounced = safeDebounce(event.onAdd);
	watcher.on("ready", async () => {
		if (event.onReady) await event.onReady();
		logger.clear();
		logger.ready("Server Ready!");
	}).on("change", async (path) => {
		logger.clear();
		logger.info(`${path} changed`);
		await onChangeDebounced(path);
	}).on("add", async (path, stats) => {
		if (event.onAdd) await onAddDebounced(path, stats);
	}).on("error", async (err) => {
		if (event.onError) await event.onError(err);
		logger.fail("Server start failed!");
		logger.error(err);
	});
}
function safeDebounce(fn) {
	if (!fn) return () => {};
	return debounce(async (...args) => {
		try {
			await fn(...args);
		} catch (error) {
			logger.error(error);
		}
	}, 500);
}
//#endregion
//#region src/utils/process.ts
function isRunning(query) {
	let cmd = "";
	switch (process$1.platform) {
		case "win32":
			cmd = `tasklist`;
			break;
		case "darwin":
			cmd = `ps -ax | grep ${query}`;
			break;
		case "linux":
			cmd = `ps -A`;
			break;
		default: break;
	}
	try {
		return execSync(cmd, { encoding: "utf8" }).toLowerCase().includes(query.toLowerCase());
	} catch {
		return false;
	}
}
const ExitSignals = [
	"exit",
	"SIGABRT",
	"SIGALRM",
	"SIGHUP",
	"SIGINT",
	"SIGTERM"
];
//#endregion
//#region src/utils/zotero/preference.ts
const prefsCommon = {
	"browser.dom.window.dump.enabled": true,
	"datareporting.policy.dataSubmissionEnabled": false,
	"devtools.debugger.remote-enabled": true,
	"devtools.debugger.prompt-connection": false,
	"devtools.browserconsole.contentMessages": true,
	"extensions.logging.enabled": false,
	"extensions.checkCompatibility.nightly": false,
	"extensions.update.enabled": false,
	"extensions.update.notifyUser": false,
	"extensions.enabledScopes": 5,
	"extensions.getAddons.cache.enabled": false,
	"extensions.installDistroAddons": false,
	"extensions.autoDisableScopes": 10,
	"app.update.enabled": false,
	"xpinstall.signatures.required": false,
	"browser.link.open_newwindow": 3
};
const prefsFirefox = {
	"browser.startup.homepage": "about:blank",
	"startup.homepage_welcome_url": "about:blank",
	"startup.homepage_welcome_url.additional": "",
	"devtools.errorconsole.enabled": true,
	"devtools.chrome.enabled": true,
	"urlclassifier.updateinterval": 172800,
	"browser.safebrowsing.provider.0.gethashURL": "http://localhost/safebrowsing-dummy/gethash",
	"browser.safebrowsing.provider.0.keyURL": "http://localhost/safebrowsing-dummy/newkey",
	"browser.safebrowsing.provider.0.updateURL": "http://localhost/safebrowsing-dummy/update",
	"browser.selfsupport.url": "https://localhost/selfrepair",
	"browser.reader.detectedFirstArticle": true,
	"datareporting.policy.firstRunURL": ""
};
const prefsZotero = {
	"extensions.experiments.enabled": true,
	"extensions.autoDisableScopes": 0,
	"devtools.debugger.remote-enabled": true,
	"devtools.debugger.remote-websocket": true,
	"devtools.debugger.prompt-connection": false,
	"extensions.zotero.firstRun.skipFirefoxProfileAccessCheck": true,
	"extensions.zotero.firstRunGuidance": false,
	"extensions.zotero.firstRun2": false,
	"extensions.zotero.reportTranslationFailure": false,
	"extensions.zotero.httpServer.enabled": true,
	"extensions.zotero.httpServer.port": 23124,
	"extensions.zotero.httpServer.localAPI.enabled": true,
	"extensions.zoteroMacWordIntegration.installed": true,
	"extensions.zoteroMacWordIntegration.skipInstallation": true,
	"extensions.zoteroWinWordIntegration.skipInstallation": true,
	"extensions.zoteroOpenOfficeIntegration.skipInstallation": true
};
const prefs = {
	...prefsCommon,
	...prefsFirefox,
	...prefsZotero
};
//#endregion
//#region src/utils/zotero/rdp-client.ts
const UNSOLICITED_EVENTS = new Set([
	"tabNavigated",
	"styleApplied",
	"propertyChange",
	"networkEventUpdate",
	"networkEvent",
	"propertyChange",
	"newMutations",
	"frameUpdate",
	"tabListChanged"
]);
function parseMessage(data) {
	const dataString = data.toString();
	const separatorIndex = dataString.indexOf(":");
	if (separatorIndex < 1) return { remainingData: data };
	const messageLength = Number.parseInt(dataString.substring(0, separatorIndex), 10);
	if (Number.isNaN(messageLength)) return {
		remainingData: data,
		error: /* @__PURE__ */ new Error("Error parsing RDP message length"),
		fatal: true
	};
	if (data.length - (separatorIndex + 1) < messageLength) return { remainingData: data };
	const messageContent = data.slice(separatorIndex + 1, separatorIndex + 1 + messageLength);
	const remainingData = data.slice(separatorIndex + 1 + messageLength);
	try {
		return {
			remainingData,
			parsedMessage: JSON.parse(messageContent.toString())
		};
	} catch (error) {
		return {
			remainingData,
			error,
			fatal: false
		};
	}
}
var MessagingClient = class extends EventEmitter {
	incomingData = Buffer.alloc(0);
	pendingRequests = [];
	activeRequests = /* @__PURE__ */ new Map();
	connection;
	constructor() {
		super();
	}
	async connect(port) {
		await new Promise((resolve, reject) => {
			const d = domain.create();
			d.once("error", reject);
			d.run(() => {
				const connectionOptions = {
					port,
					host: "127.0.0.1"
				};
				const conn = net.createConnection(connectionOptions, () => {
					resolve();
				});
				this.connection = conn;
				conn.on("data", this.onData.bind(this));
				conn.on("error", reject);
				conn.on("end", this.onEnd.bind(this));
				conn.on("timeout", this.onTimeout.bind(this));
				this.expectReply("root", {
					resolve,
					reject
				});
			});
		});
	}
	disconnect() {
		if (!this.connection) return;
		this.connection.removeAllListeners();
		this.connection.end();
		this.rejectAllRequests(/* @__PURE__ */ new Error("RDP connection closed"));
	}
	rejectAllRequests(error) {
		this.activeRequests.forEach((deferred) => {
			deferred.reject(error);
		});
		this.activeRequests.clear();
		this.pendingRequests.forEach(({ deferred }) => {
			deferred.reject(error);
		});
		this.pendingRequests = [];
	}
	async request(requestProps) {
		const request = typeof requestProps === "string" ? {
			to: "root",
			type: requestProps
		} : requestProps;
		if (!request.to) throw new Error(`Unexpected RDP request without target actor: ${request.type}`);
		return await new Promise((resolve, reject) => {
			const deferred = {
				resolve,
				reject
			};
			this.pendingRequests.push({
				request,
				deferred
			});
			this.flushPendingRequests();
		});
	}
	flushPendingRequests() {
		this.pendingRequests = this.pendingRequests.filter(({ request, deferred }) => {
			if (this.activeRequests.has(request.to)) return true;
			if (!this.connection) throw new Error("RDP connection closed");
			try {
				const messageString = `${Buffer.from(JSON.stringify(request)).length}:${JSON.stringify(request)}`;
				this.connection.write(messageString);
				this.expectReply(request.to, deferred);
			} catch (err) {
				deferred.reject(err);
			}
			return false;
		});
	}
	expectReply(targetActor, deferred) {
		if (this.activeRequests.has(targetActor)) throw new Error(`${targetActor} does already have an active request`);
		this.activeRequests.set(targetActor, deferred);
	}
	onData(data) {
		this.incomingData = Buffer.concat([this.incomingData, data]);
		while (this.readMessage());
	}
	readMessage() {
		const { remainingData, parsedMessage, error, fatal } = parseMessage(this.incomingData);
		this.incomingData = remainingData;
		if (error) {
			this.emit("error", /* @__PURE__ */ new Error(`Error parsing RDP packet: ${String(error)}`));
			if (fatal) this.disconnect();
			return !fatal;
		}
		if (!parsedMessage) return false;
		this.handleMessage(parsedMessage);
		return true;
	}
	handleMessage(message) {
		if (!message.from) {
			if (message.error) {
				this.emit("rdp-error", message);
				return;
			}
			this.emit("error", /* @__PURE__ */ new Error(`Received an RDP message without a sender actor: ${JSON.stringify(message)}`));
			return;
		}
		if (message.type && UNSOLICITED_EVENTS.has(message.type)) {
			this.emit("unsolicited-event", message);
			return;
		}
		if (this.activeRequests.has(message.from)) {
			const deferred = this.activeRequests.get(message.from);
			this.activeRequests.delete(message.from);
			if (message.error) deferred?.reject(message);
			else deferred?.resolve(message);
			this.flushPendingRequests();
			return;
		}
		if (message.type === "addonListChanged") return;
		this.emit("error", /* @__PURE__ */ new Error(`Received unexpected message:\n ${JSON.stringify(message)}`));
	}
	onError(error) {
		this.emit("error", error);
	}
	onEnd() {
		this.emit("end");
	}
	onTimeout() {
		this.emit("timeout");
	}
};
//#endregion
//#region src/utils/zotero/remote-zotero.ts
const MAX_RETRIES = 150;
const RETRY_INTERVAL = 1e3;
function requestErrorToMessage(err) {
	if (err instanceof Error) return String(err);
	return `${err.error}: ${err.message}`;
}
function isErrorWithCode(codeWanted, error) {
	if (Array.isArray(codeWanted) && codeWanted.includes(error.code)) return true;
	else if (error.code === codeWanted) return true;
	return false;
}
var RemoteFirefox = class {
	client;
	checkedForAddonReloading;
	constructor() {
		this.client = new MessagingClient();
		this.checkedForAddonReloading = false;
		this.client.on("disconnect", () => {
			logger.debug("Received \"disconnect\" from Firefox client");
		});
		this.client.on("end", () => {
			logger.debug("Received \"end\" from Firefox client");
		});
		this.client.on("unsolicited-event", (info) => {
			logger.debug(`Received message from client: ${JSON.stringify(info)}`);
		});
		this.client.on("rdp-error", (rdpError) => {
			logger.debug(`Received error from client: ${JSON.stringify(rdpError)}`);
		});
		this.client.on("error", (error) => {
			logger.debug(`Received error from client: ${String(error)}`);
		});
	}
	async connect(port) {
		let lastError;
		for (const _ of Array.from({ length: MAX_RETRIES })) try {
			this.client = new MessagingClient();
			await this.client.connect(port);
			return this.client;
		} catch (error) {
			logger.debug("Failed to connecte to RDP client, retry...");
			if (isErrorWithCode("ECONNREFUSED", error)) {
				await new Promise((resolve) => setTimeout(resolve, RETRY_INTERVAL));
				lastError = error;
			} else throw error;
		}
		logger.error(`Unable to connect to Zotero. Too many retries.`);
		throw lastError;
	}
	disconnect() {
		this.client.disconnect();
	}
	async addonRequest(addon, request) {
		try {
			return await this.client.request({
				to: addon.actor,
				type: request
			});
		} catch (err) {
			logger.debug(`Client responded to '${request}' request with error: ${err}`);
			const message = requestErrorToMessage(err);
			throw new Error(`Remote Firefox: addonRequest() error: ${message}`);
		}
	}
	async getAddonsActor() {
		try {
			const response = await this.client.request("getRoot");
			if (response.addonsActor == null) return Promise.reject(/* @__PURE__ */ new Error("This version of Firefox does not provide an add-ons actor for remote installation."));
			return response.addonsActor;
		} catch (err) {
			logger.debug(`Falling back to listTabs because getRoot failed", ${err}`);
		}
		try {
			const response = await this.client.request("listTabs");
			if (response.addonsActor == null) {
				logger.debug(`listTabs returned a falsey addonsActor: ${JSON.stringify(response)}`);
				return Promise.reject(/* @__PURE__ */ new Error("This is an older version of Firefox that does not provide an add-ons actor for remote installation. Try Firefox 49 or higher."));
			}
			return response.addonsActor;
		} catch (err) {
			logger.debug(`listTabs error: ${err}`);
			const message = requestErrorToMessage(err);
			throw new Error(`Remote Firefox: listTabs() error: ${message}`);
		}
	}
	async installTemporaryAddon(addonPath) {
		const addonsActor = await this.getAddonsActor();
		try {
			const response = await this.client.request({
				to: addonsActor,
				type: "installTemporaryAddon",
				addonPath
			});
			logger.debug(`installTemporaryAddon: ${JSON.stringify(response)}`);
			logger.info(`Installed ${addonPath} as a temporary add-on`);
			return response;
		} catch (err) {
			const message = requestErrorToMessage(err);
			throw new Error(`installTemporaryAddon: Error: ${message}`);
		}
	}
	async getInstalledAddon(addonId) {
		try {
			const response = await this.client.request("listAddons");
			for (const addon of response.addons) if (addon.id === addonId) return addon;
			logger.debug(`Remote Firefox has these addons: ${response.addons.map((a) => a.id)}`);
			return Promise.reject(/* @__PURE__ */ new Error("The remote Firefox does not have your extension installed"));
		} catch (err) {
			const message = requestErrorToMessage(err);
			throw new Error(`Remote Firefox: listAddons() error: ${message}`);
		}
	}
	async checkForAddonReloading(addon) {
		if (this.checkedForAddonReloading) return addon;
		else {
			/**
			* {
			*   "from": "server1.conn0.webExtensionDescriptor8",
			*   "requestTypes": [
			*     "reload",
			*     "terminateBackgroundScript",
			*     "connect",
			*     "getTarget",
			*     "reloadDescriptor",
			*     "getWatcher"
			*   ]
			* }
			*
			*/
			const response = await this.addonRequest(addon, "requestTypes");
			if (!response.requestTypes.includes("reload")) {
				const supportedRequestTypes = JSON.stringify(response.requestTypes);
				logger.debug(`Remote Firefox only supports: ${supportedRequestTypes}`);
				throw new Error("This Firefox version does not support add-on reloading. Re-run with --no-reload");
			} else {
				this.checkedForAddonReloading = true;
				return addon;
			}
		}
	}
	async reloadAddon(addonId) {
		const addon = await this.getInstalledAddon(addonId);
		await this.checkForAddonReloading(addon);
		await this.addonRequest(addon, "reload");
		logger.success(`Last extension reload: ${(/* @__PURE__ */ new Date()).toTimeString()}`);
	}
};
function findFreeTcpPort() {
	return new Promise((resolve) => {
		const srv = net.createServer();
		srv.listen(0, "127.0.0.1", () => {
			const addr = srv.address();
			if (addr && typeof addr !== "string") {
				const freeTcpPort = addr.port;
				srv.close(() => resolve(freeTcpPort));
			}
		});
	});
}
//#endregion
//#region src/utils/zotero-runner.ts
const default_options = {
	binary: {
		args: [],
		devtools: true
	},
	profile: {
		path: "./.scaffold/profile",
		dataDir: "",
		createIfMissing: true,
		customPrefs: {}
	},
	plugins: {
		asProxy: false,
		list: []
	}
};
var ZoteroRunner = class {
	options;
	remoteFirefox = new RemoteFirefox();
	zotero;
	constructor(options) {
		this.options = toMerged(default_options, options);
		if (!options.binary.path) throw new Error("Binary path must be provided.");
		if (!options.profile.path && !options.profile.dataDir) this.options.profile.dataDir = "./.scaffold/data";
		logger.debug(this.options);
	}
	get default_profile_path() {
		return default_options.profile.path;
	}
	async run() {
		await this.setupProfile();
		await this.startZoteroInstance();
		if (!this.options.plugins.asProxy) await this.installTemporaryPlugins();
	}
	/**
	* Preparing the development environment
	*
	* When asProxy=true, generate a proxy file and replace prefs.
	*
	* @see https://www.zotero.org/support/dev/client_coding/plugin_development#setting_up_a_plugin_development_environment
	*/
	async setupProfile() {
		const { path, createIfMissing } = this.options.profile;
		if (!await pathExists(path)) if (createIfMissing) await this.createProfile(this.default_profile_path);
		else throw new Error("The 'profile.path' must be provided when 'createIfMissing' is false.");
		const prefsPath = join(this.options.profile.path, "prefs.js");
		const prefsManager = new PrefsManager("user_pref");
		prefsManager.setPrefs(prefs);
		if (await pathExists(prefsPath)) await prefsManager.read(prefsPath);
		prefsManager.setPrefs(this.options.profile.customPrefs);
		prefsManager.setPrefs({
			"extensions.lastAppBuildId": null,
			"extensions.lastAppVersion": null
		});
		await prefsManager.write(prefsPath);
		if (this.options.plugins.asProxy) await this.installProxyPlugins();
	}
	async createProfile(path) {
		logger.debug(`Creating profile at ${this.default_profile_path}...`);
		await ensureDir(path);
	}
	async startZoteroInstance() {
		let args = ["--purgecaches", "no-remote"];
		if (this.options.profile.path) args.push("-profile", resolve(this.options.profile.path));
		if (this.options.profile.dataDir) args.push("--dataDir", resolve(this.options.profile.dataDir));
		if (this.options.binary.devtools) args.push("--jsdebugger");
		if (this.options.binary.args) args = [...args, ...this.options.binary.args];
		const remotePort = await findFreeTcpPort();
		args.push("-start-debugger-server", String(remotePort));
		logger.debug(`Zotero start args: ${args}`);
		const env = {
			...process$1.env,
			XPCOM_DEBUG_BREAK: "stack",
			NS_TRACE_MALLOC_DISABLE_STACKS: "1"
		};
		if (!await pathExists(this.options.binary.path)) throw new Error("The Zotero binary not found.");
		this.zotero = spawn(this.options.binary.path, args, { env });
		logger.debug(`Zotero started, pid: ${this.zotero.pid}`);
		this.zotero.stdout?.on("data", (_data) => {});
		logger.debug("Connecting to the remote Firefox debugger...");
		await this.remoteFirefox.connect(remotePort);
		logger.debug(`Connected to the remote Firefox debugger on port: ${remotePort}`);
	}
	async installTemporaryPlugins() {
		for (const plugin of this.options.plugins.list) if (!await this.remoteFirefox.installTemporaryAddon(resolve(plugin.sourceDir)).then((installResult) => {
			return installResult.addon.id;
		})) throw new Error("Unexpected missing addonId in the installAsTemporaryAddon result");
	}
	async installProxyPlugin(id, sourceDir) {
		const addonProxyFilePath = join(this.options.profile.path, `extensions/${id}`);
		const buildPath = resolve(sourceDir);
		await outputFile(addonProxyFilePath, buildPath);
		logger.debug([
			`Addon proxy file has been updated.`,
			`  File path: ${addonProxyFilePath}`,
			`  Addon path: ${buildPath}`
		].join("\n"));
		const addonXpiFilePath = join(this.options.profile.path, `extensions/${id}.xpi`);
		if (await pathExists(addonXpiFilePath)) {
			await remove(addonXpiFilePath);
			logger.debug(`XPI file found, removed.`);
		}
		const addonInfoFilePath = join(this.options.profile.path, "extensions.json");
		if (await pathExists(addonInfoFilePath)) {
			const content = await readJSON(addonInfoFilePath);
			content.addons = content.addons.map((addon) => {
				if (addon.id === id && addon.active === false) {
					addon.active = true;
					addon.userDisabled = false;
					logger.debug(`Active plugin ${id} in extensions.json.`);
				}
				return addon;
			});
			await outputJSON(addonInfoFilePath, content);
		}
	}
	async installProxyPlugins() {
		for (const { id, sourceDir } of this.options.plugins.list) await this.installProxyPlugin(id, sourceDir);
	}
	async reloadTemporaryPluginById(id) {
		await this.remoteFirefox.reloadAddon(id);
	}
	async reloadTemporaryPluginBySourceDir(sourceDir) {
		const addonId = this.options.plugins.list.find((p) => p.sourceDir === sourceDir)?.id;
		if (!addonId) return {
			sourceDir,
			reloadError: /* @__PURE__ */ new Error(`Extension not reloadable: no addonId has been mapped to "${sourceDir}"`)
		};
		try {
			await this.remoteFirefox.reloadAddon(addonId);
		} catch (error) {
			return {
				sourceDir,
				reloadError: error
			};
		}
		return {
			sourceDir,
			reloadError: void 0
		};
	}
	async reloadAllTemporaryPlugins() {
		for (const { sourceDir } of this.options.plugins.list) {
			const res = await this.reloadTemporaryPluginBySourceDir(sourceDir);
			if (res.reloadError instanceof Error) logger.error(res.reloadError);
		}
	}
	async reloadProxyPluginByZToolkit(id, name, version) {
		const reloadScript = `
    (async () => {
    Services.obs.notifyObservers(null, "startupcache-invalidate", null);
    const { AddonManager } = ChromeUtils.import("resource://gre/modules/AddonManager.jsm");
    const addon = await AddonManager.getAddonByID("${id}");
    await addon.reload();
    const progressWindow = new Zotero.ProgressWindow({ closeOnClick: true });
    progressWindow.changeHeadline("${name} Hot Reload");
    progressWindow.progress = new progressWindow.ItemProgress(
        "chrome://zotero/skin/tick.png",
        "VERSION=${version}, BUILD=${(/* @__PURE__ */ new Date()).toLocaleString()}. By zotero-plugin-toolkit"
    );
    progressWindow.progress.setProgress(100);
    progressWindow.show();
    progressWindow.startCloseTimer(5000);
    })()`;
		const url = `zotero://ztoolkit-debug/?run=${encodeURIComponent(reloadScript)}`;
		execSync(`${`"${this.options.binary.path}" --purgecaches -profile "${this.options.profile.path}"`} -url "${url}"`);
	}
	async reloadAllProxyPlugins() {
		for (const { id } of this.options.plugins.list) {
			await this.reloadProxyPluginByZToolkit(id, id, id);
			await delay(2e3);
		}
	}
	async reloadAllPlugins() {
		if (this.options.plugins.asProxy) await this.reloadAllProxyPlugins();
		else await this.reloadAllTemporaryPlugins();
	}
	exit() {
		this.zotero?.kill();
		killZotero();
	}
};
function killZotero() {
	function kill() {
		try {
			if (process$1.env.ZOTERO_PLUGIN_KILL_COMMAND) execSync(process$1.env.ZOTERO_PLUGIN_KILL_COMMAND);
			else if (isWindows) execSync("taskkill /f /im zotero.exe");
			else if (isMacOS) execSync("kill -9 $(ps -x | grep zotero)");
			else if (isLinux) execSync("pkill -9 zotero");
			else logger.error("No commands found for this operating system.");
		} catch {
			logger.fail("Kill Zotero failed.");
		}
	}
	if (isRunning("zotero")) kill();
	else logger.fail("No Zotero instance is currently running.");
}
//#endregion
//#region src/core/server.ts
var Serve = class extends Base {
	builder;
	runner;
	_zoteroBinPath;
	constructor(ctx) {
		super(ctx);
		process$1.env.NODE_ENV ??= "development";
		this.builder = new Build(ctx);
	}
	async run() {
		this.runner = new ZoteroRunner({
			binary: {
				path: this.zoteroBinPath,
				devtools: this.ctx.server.devtools,
				args: this.ctx.server.startArgs
			},
			profile: {
				path: this.profilePath,
				dataDir: this.dataDir,
				createIfMissing: this.ctx.server.createProfileIfMissing,
				customPrefs: this.ctx.server.prefs
			},
			plugins: {
				list: [{
					id: this.ctx.id,
					sourceDir: join(this.ctx.dist, "addon")
				}],
				asProxy: this.ctx.server.asProxy
			}
		});
		await this.ctx.hooks.callHook("serve:init", this.ctx);
		if (this.ctx.server.prebuild) {
			await this.builder.run();
			await this.ctx.hooks.callHook("serve:prebuild", this.ctx);
		}
		await this.runner.run();
		this.runner.zotero?.on("exit", this.onZoteroExit);
		this.runner.zotero?.on("close", this.onZoteroExit);
		await this.watch();
	}
	/**
	* watch source dir and build when file changed
	*/
	async watch() {
		const { source, watchIgnore } = this.ctx;
		watch(source, watchIgnore, {
			onReady: async () => {
				await this.ctx.hooks.callHook("serve:ready", this.ctx);
			},
			onChange: async (path) => {
				await this.ctx.hooks.callHook("serve:onChanged", this.ctx, path);
				if (path.endsWith(".ts") || path.endsWith(".tsx")) await this.builder.bundle();
				else await this.builder.run();
				await this.reload();
			}
		});
	}
	async reload() {
		this.logger.tip("Reloading...");
		await this.runner?.reloadAllPlugins();
		await this.ctx.hooks.callHook("serve:onReloaded", this.ctx);
	}
	exit = () => {
		this.logger.info("Server shutdown by user request.");
		this.runner?.exit();
		this.ctx.hooks.callHook("serve:exit", this.ctx);
		process$1.exit();
	};
	onZoteroExit = (_code, _signal) => {
		this.logger.info(`Zotero terminated.`);
		process$1.exit();
	};
	get zoteroBinPath() {
		if (this._zoteroBinPath) return this._zoteroBinPath;
		this._zoteroBinPath = process$1.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH;
		if (!this._zoteroBinPath || !existsSync(this._zoteroBinPath)) throw new Error("The Zotero binary not found.");
		return this._zoteroBinPath;
	}
	get profilePath() {
		return process$1.env.ZOTERO_PLUGIN_PROFILE_PATH;
	}
	get dataDir() {
		return process$1.env.ZOTERO_PLUGIN_DATA_DIR;
	}
};
//#endregion
//#region src/constant.ts
const CACHE_DIR = `.scaffold/cache`;
const TESTER_PROFILE_DIR = `.scaffold/test/profile`;
const TESTER_DATA_DIR = `.scaffold/test/data`;
const TESTER_PLUGIN_DIR = `.scaffold/test/resource`;
const TESTER_PLUGIN_REF = "zotero-plugin-scaffold-test-runner";
const TESTER_PLUGIN_ID = "scaffold-test@northword.cn";
//#endregion
//#region src/core/tester/headless.ts
async function prepareHeadless() {
	await installXvfb();
	await installDepsForUbuntu24();
	await installZoteroLinux();
	await new Xvfb({ timeout: 2e3 }).start();
}
function isPackageInstalled(packageName) {
	try {
		execSync(`dpkg-query -W ${packageName}`, { stdio: "ignore" });
		return true;
	} catch {
		return false;
	}
}
function installPackage(packageName) {
	const debug = isDebug || logger.level <= LOG_LEVEL.DEBUG;
	try {
		logger.debug(`Installing ${packageName}...`);
		execSync(`sudo apt update && sudo apt install -y ${packageName}`, { stdio: debug ? "inherit" : "pipe" });
		logger.debug(`${packageName} installed successfully.`);
	} catch (error) {
		logger.fail(`Failed to install ${packageName}. ${error}`);
		throw error;
	}
}
function checkAndInstallDependencies(packages) {
	packages.forEach((pkg) => {
		if (isPackageInstalled(pkg)) logger.debug(`${pkg} is already installed.`);
		else installPackage(pkg);
	});
}
async function installXvfb() {
	if (!isLinux) {
		logger.error("Unsupported platform. Please install Xvfb manually.");
		return;
	}
	const osId = execSync("cat /etc/os-release | grep '^ID='").toString();
	if (!(osId.includes("ubuntu") || osId.includes("debian"))) {
		logger.error("Unsupported Linux distribution.");
		return;
	}
	checkAndInstallDependencies([
		"xvfb",
		"x11-xkb-utils",
		"xkb-data"
	]);
}
async function installDepsForUbuntu24() {
	if (!execSync("cat /etc/os-release | grep '^VERSION_ID='").toString().includes("24")) {
		logger.error("Skip to install deps due to version not 24.04.");
		return;
	}
	checkAndInstallDependencies(["libasound2t64", "libdbus-glib-1-2"]);
}
async function installZoteroLinux() {
	if (process$1.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH) {
		logger.debug("Local Zotero found, skip to download.");
		return;
	}
	logger.debug("Installing Zotero...");
	execSync(`cd ${CACHE_DIR}`);
	execSync("wget -O zotero.tar.bz2 'https://www.zotero.org/download/client/dl?platform=linux-x86_64&channel=beta'", { stdio: "pipe" });
	execSync("tar -xvf zotero.tar.bz2", { stdio: "pipe" });
	process$1.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH = `${process$1.cwd()}/Zotero_linux-x86_64/zotero`;
	execSync("cd -");
}
//#endregion
//#region src/core/tester/http-reporter.ts
var TestHttpReporter = class {
	_server;
	_port;
	passed = 0;
	failed = 0;
	constructor() {}
	async getPort() {
		this._port = await findFreeTcpPort();
	}
	get port() {
		return this._port;
	}
	async start() {
		if (!this._port) await this.getPort();
		this._server = http.createServer(this.handleRequest.bind(this));
		this._server.listen(this._port, () => {
			logger.debug(`Server is listening on http://localhost:${this.port}`);
		});
		return this;
	}
	handleRequest(req, res) {
		if (req.method === "GET" && req.url === "/") {
			res.writeHead(200, { "Content-Type": "text/plain" });
			res.end("Zotero Plugin Test Server is running");
		} else if (req.method === "POST" && req.url === "/update") {
			let body = "";
			req.on("data", (chunk) => {
				body += chunk;
			});
			req.on("end", async () => {
				try {
					const jsonData = JSON.parse(body);
					await this.onData(jsonData);
					res.writeHead(200, { "Content-Type": "application/json" });
					res.end(JSON.stringify({ message: "Results received successfully" }));
				} catch (error) {
					logger.error(`Error parsing JSON:, ${error}`);
					res.writeHead(400, { "Content-Type": "application/json" });
					res.end(JSON.stringify({ error: "Invalid JSON" }));
				}
			});
		} else {
			res.writeHead(404, { "Content-Type": "application/json" });
			res.end(JSON.stringify({ error: "Not Found" }));
		}
	}
	async onData(body) {
		const { type, data } = body;
		const logger_option = { space: data.indents - 1 };
		switch (type) {
			case "debug":
				logger.log(data);
				break;
			case "start":
				logger.newLine();
				break;
			case "suite":
				if (data.title) logger.tip(data.title, logger_option);
				break;
			case "pass":
				logger.success(`${data.title} ${styleText("gray", `${data.duration}ms`)}`, logger_option);
				break;
			case "fail": {
				logger.fail(styleText("red", `${data.title}, ${data?.error?.message}`), logger_option);
				let expected = data.error.expected;
				let received = data.error.actual;
				function formatOutput(obj, indent) {
					const lines = JSON.stringify(obj, null, 2).split("\n");
					return `${lines[0]}\n${lines.slice(1).map((line) => "  ".repeat(indent) + line).join("\n")}`;
				}
				if (expected && typeof expected === "object") expected = formatOutput(expected, data.indents + 1);
				if (received && typeof received === "object") received = formatOutput(received, data.indents + 1);
				logger.log(`Expected: ${styleText("green", String(expected))}`, { space: data.indents + .5 });
				logger.log(`Received: ${styleText("red", String(received))}`, { space: data.indents + .5 });
				break;
			}
			case "pending":
				logger.info(`${data.title} pending`, logger_option);
				break;
			case "suite end": break;
			case "end":
				this.passed = data.passed;
				this.failed = data.failed;
				logger.newLine();
				if (this.failed === 0) logger.success(`Test run completed - ${this.passed} passed`);
				else logger.fail(`Test run completed - ${this.passed} passed, ${this.failed} failed`);
				break;
			case "error":
				logger.error(body.data?.title);
				this.failed++;
				break;
			default:
				logger.log(data);
				break;
		}
	}
	stop() {
		this._server?.close();
	}
};
//#endregion
//#region src/utils/file.ts
async function saveResource(url, path) {
	await outputFile(path, await (await fetch(url)).text());
}
//#endregion
//#region src/core/tester/test-bundler-template/raw/bootstrap.js?raw
var bootstrap_default = "/* eslint-disable eslint-comments/no-unlimited-disable */\n/* eslint-disable */\n// Code generated by the zotero-plugin-scaffold tester\nvar chromeHandle;\n\nfunction install(data, reason) {}\n\nasync function startup({ id, version, resourceURI, rootURI }, reason) {\n  await Zotero.initializationPromise;\n  const aomStartup = Components.classes[\n    \"@mozilla.org/addons/addon-manager-startup;1\"\n  ].getService(Components.interfaces.amIAddonManagerStartup);\n  const manifestURI = Services.io.newURI(rootURI + \"manifest.json\");\n  chromeHandle = aomStartup.registerChrome(manifestURI, [\n    [\"content\", \"__TESTER_PLUGIN_REF__\", rootURI + \"content/\"],\n  ]);\n\n  launchTests().catch((error) => {\n    Zotero.debug(error);\n    Zotero.HTTP.request(\n      \"POST\",\n      \"http://localhost:__PORT__/update\",\n      {\n        body: JSON.stringify({\n          type: \"error\",\n          data: {\n            title: error.message,\n          },\n        }),\n      }\n    );\n    // Although this is an exception, we still need to use the status code 0, \n    // because status code 1 will cause Zotero to auto-restart.\n    Zotero.Utilities.Internal.quit(0);\n  });\n}\n\nfunction onMainWindowLoad({ window: win }) {}\n\nfunction onMainWindowUnload({ window: win }) {}\n\nfunction shutdown({ id, version, resourceURI, rootURI }, reason) {\n  if (reason === APP_SHUTDOWN) {\n    return;\n  }\n\n  if (chromeHandle) {\n    chromeHandle.destruct();\n    chromeHandle = null;\n  }\n}\n\nfunction uninstall(data, reason) {}\n\nasync function launchTests() {\n  // Delay to allow plugin to fully load before opening the test page\n  if (__STARTUP_DELAY__) {\n    await Zotero.Promise.delay(__STARTUP_DELAY__);\n  }\n\n  const waitForPlugin = \"__WAIT_FOR_PLUGIN__\";\n  if (waitForPlugin) {\n    // Wait for a plugin to be installed\n    await waitUtilAsync(() => {\n      try {\n        return !!eval(waitForPlugin)();\n      } catch (error) {\n        return false;\n      }\n    }).catch(() => {\n      throw new Error(\"Internal: Plugin awaiting timeout\");\n    });\n  }\n\n  Services.ww.openWindow(\n    null,\n    \"chrome://__TESTER_PLUGIN_REF__/content/index.xhtml\",\n    \"Zotero Plugin Scaffold Test Runnner\",\n    \"chrome,centerscreen,resizable=yes\",\n    {}\n  );\n}\n\nfunction waitUtilAsync(condition, interval = 100, timeout = 1e4) {\n  return new Promise((resolve, reject) => {\n    const start = Date.now();\n    const intervalId = setInterval(() => {\n      if (condition()) {\n        clearInterval(intervalId);\n        resolve();\n      } else if (Date.now() - start > timeout) {\n        clearInterval(intervalId);\n        reject();\n      }\n    }, interval);\n  });\n}\n";
//#endregion
//#region src/core/tester/test-bundler-template/raw/index.html?raw
var raw_default = "<!-- Generated by zotero-plugin-scaffold -->\n<html lang=\"en\" xmlns=\"http://www.w3.org/1999/xhtml\">\n  <head>\n    <meta charset=\"UTF-8\" />\n    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\" />\n    <title>Zotero Plugin Test</title>\n    <style>\n      html {\n        min-width: 400px;\n        min-height: 600px;\n      }\n      body {\n        font-family: Arial, sans-serif;\n      }\n    </style>\n  </head>\n  <body>\n    <div id=\"mocha\"></div>\n\n    <!-- Include Zotero Vars -->\n    <script src=\"chrome://zotero/content/include.js\"><\/script>\n\n    <!-- Mocha and Chai Libraries -->\n    <script src=\"mocha.js\"><\/script>\n    <script src=\"chai.js\"><\/script>\n\n    <!-- Setup Mocha -->\n    <script>\n      __SETUP_CODE__;\n    <\/script>\n\n    <!-- Unit tests -->\n    __TEST_FILES__\n\n    <!-- Run Mocha -->\n    <script class=\"mocha-exec\">\n      mocha.run();\n    <\/script>\n  </body>\n</html>\n";
//#endregion
//#region src/core/tester/test-bundler-template/raw/manifest.json?raw
var manifest_default = "{\n  \"manifest_version\": 2,\n  \"name\": \"Zotero Plugin Scaffold Test Runner\",\n  \"version\": \"0.0.1\",\n  \"description\": \"Test suite for the Zotero plugin. This is a runtime-generated plugin only for testing purposes.\",\n  \"applications\": {\n    \"zotero\": {\n      \"id\": \"__TESTER_PLUGIN_ID__\",\n      \"update_url\": \"https://example.com\",\n      \"strict_max_version\": \"999.*.*\"\n    }\n  }\n}\n";
//#endregion
//#region src/core/tester/test-bundler-template/raw/mocha-setup.js?raw
var mocha_setup_default = "/* eslint-disable eslint-comments/no-unlimited-disable */\n/* eslint-disable */\n// Generated by zotero-plugin-scaffold\nmocha.setup({ \n  ui: \"bdd\",\n  reporter: Reporter,\n  timeout: __TIMEOUT__ || 10000,\n  bail: __ABORT_ON_FAIL__\n});\n\nwindow.expect = chai.expect;\nwindow.assert = chai.assert;\n\nasync function send(data) {\n  const req = await Zotero.HTTP.request(\n    \"POST\",\n    \"http://localhost:__PORT__/update\",\n    {\n      body: JSON.stringify(data),\n    }\n  );\n\n  if (req.status !== 200) {\n    dump(\"Error sending data to server\" + req.responseText);\n    return null;\n  } else {\n    const result = JSON.parse(req.responseText);\n    return result;\n  }\n}\n\nwindow.debug = function (data) {\n  send({ type: \"debug\", data });\n};\n\nlet indents = 0,\n  passed = 0,\n  failed = 0,\n  aborted = false;\n\n// Inherit the default test settings from Zotero\nfunction Reporter(runner) {\n  function indent() {\n    return \"  \".repeat(indents);\n  }\n\n  function dump(str) {\n    document.querySelector(\"#mocha\").innerText += str;\n  }\n\n  runner.on(\"start\", async function () {\n    console.log(\"start\")\n    await send({ type: \"start\", data: { indents } });\n  });\n\n  runner.on(\"suite\", async function (suite) {\n    console.log(\"suite\", suite)\n    indents++;\n    const str = indent() + suite.title + \"\\n\";\n    dump(str);\n    await send({ type: \"suite\", data: { title: suite.title, root: suite.root, indents } });\n  });\n\n  runner.on(\"suite end\", async function (suite) {\n    console.log(\"suite end\", suite)\n    indents--;\n    const str = indents === 1 ? \"\\n\" : \"\";\n    dump(str);\n    await send({ type: \"suite end\", data: { title: suite.title, root: suite.root, indents } });\n  });\n\n  runner.on(\"pending\", async function (test) {\n    console.log(\"pending\", test)\n    const str = indent() + \"pending  -\" + test.title + \"\\n\";\n    dump(str);\n    await send({ type: \"pending\", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, indents: indents + 1 } });\n  });\n\n  runner.on(\"pass\", async function (test) {\n    console.log(\"pass\", test)\n    passed++;\n    let str = indent() + Mocha.reporters.Base.symbols.ok + \" \" + test.title;\n    if (\"fast\" != test.speed) {\n      str += \" (\" + Math.round(test.duration) + \" ms)\";\n    }\n    str += \"\\n\";\n    dump(str);\n    await send({ type: \"pass\", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, indents: indents + 1 } });\n\n  });\n\n  runner.on(\"fail\", async function (test, error) {\n    console.log(\"fail\", test, error)\n\n    failed++;\n    let indentStr = indent();\n    const str =\n      indentStr +\n      Mocha.reporters.Base.symbols.err +\n      (Zotero.automatedTest ? \"\" : \"\\x07\") +\n      \" \" +\n      test.title +\n      \"\\n\" +\n      indentStr +\n      \"  \" +\n      error.message +\n      \"\\n\\n\";\n    dump(str);\n\n    await send({ type: \"fail\", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, error, indents: indents + 1 } });\n\n  });\n\n  runner.on(\"end\", async function () {\n    console.log(\"end\")\n    const str =\n      passed +\n      \"/\" +\n      (passed + failed) +\n      \" tests passed\" +\n      (aborted ? \" -- aborting\" : \"\") +\n      \"\\n\";\n    dump(str);\n\n    await send({\n      type: \"end\",\n      data: { passed: passed, failed: failed, aborted: aborted, str, indents },\n    });\n\n    // Must exit on Zotero side, otherwise the exit code will not be 0 and CI will fail\n    if (__EXIT_ON_FINISH__) {\n      Zotero.Utilities.Internal.quit(0);\n    }\n  });\n}\n";
//#endregion
//#region src/core/tester/test-bundler-template/index.ts
function generateManifest() {
	const manifestStr = manifest_default.replaceAll("__TESTER_PLUGIN_ID__", TESTER_PLUGIN_ID);
	return JSON.parse(manifestStr);
}
function generateBootstrap(options) {
	return bootstrap_default.replaceAll("__PORT__", String(options.port)).replaceAll("__STARTUP_DELAY__", String(options.startupDelay || 1e3)).replaceAll("__WAIT_FOR_PLUGIN__", options.waitForPlugin).replaceAll("__TESTER_PLUGIN_REF__", TESTER_PLUGIN_REF);
}
function generateHtml(setupCode, testFiles) {
	const tests = testFiles.map((f) => `<script src="${f}"><\/script>`).join("\n    ");
	return raw_default.replaceAll("__TEST_FILES__", tests).replaceAll("__SETUP_CODE__", setupCode);
}
function generateMochaSetup(options) {
	return mocha_setup_default.replaceAll("__TIMEOUT__", String(options.timeout || 1e4)).replaceAll("__PORT__", String(options.port)).replaceAll("__ABORT_ON_FAIL__", String(options.abortOnFail)).replaceAll("__EXIT_ON_FINISH__", String(options.exitOnFinish ? "true" : "false"));
}
//#endregion
//#region src/core/tester/test-bundler.ts
var TestBundler = class {
	esbuildContext;
	constructor(ctx, port) {
		this.ctx = ctx;
		this.port = port;
	}
	async generate() {
		await this.generateTestResources();
		await this.createTestHtml();
	}
	async regenerate(changedFile) {
		const tests = findImpactedTests(changedFile, (await this.esbuildContext?.rebuild())?.metafile);
		await this.createTestHtml(tests);
	}
	async generateTestResources() {
		const manifest = generateManifest();
		await outputJSON(`${TESTER_PLUGIN_DIR}/manifest.json`, manifest, { spaces: 2 });
		const bootstrap = generateBootstrap({
			port: this.port,
			startupDelay: this.ctx.test.startupDelay,
			waitForPlugin: this.ctx.test.waitForPlugin
		});
		await outputFile(`${TESTER_PLUGIN_DIR}/bootstrap.js`, bootstrap);
		await this.copyTestLibraries();
		await this.bundleTests();
	}
	async copyTestLibraries() {
		await Promise.all([{
			name: "mocha.js",
			local: "node_modules/mocha/mocha.js",
			remote: "https://cdn.jsdelivr.net/npm/mocha/mocha.js"
		}, {
			name: "chai.js",
			local: "",
			remote: "https://www.chaijs.com/chai.js"
		}].map(async (pkg) => {
			const targetPath = `${TESTER_PLUGIN_DIR}/content/${pkg.name}`;
			if (pkg.local && await pathExists(pkg.local)) {
				logger.debug(`Local ${pkg.name} package found`);
				await copy(pkg.local, targetPath);
				return;
			}
			const cachePath = `${CACHE_DIR}/${pkg.name}`;
			if (await pathExists(`${cachePath}`)) {
				logger.debug(`Cache ${pkg.name} package found`);
				await copy(cachePath, targetPath);
				return;
			}
			logger.info(`No local ${pkg.name} found, we recommend you install ${pkg.name} package locally.`);
			await saveResource(pkg.remote, `${CACHE_DIR}/${pkg.name}`);
			await copy(cachePath, targetPath);
		}));
	}
	async bundleTests() {
		const testDirs = toArray(this.ctx.test.entries);
		this.esbuildContext = await context({
			entryPoints: (await Promise.all(testDirs.map((dir) => glob(`${dir}/**/*.{spec,test}.[jt]s`)))).flat(),
			outdir: `${TESTER_PLUGIN_DIR}/content/units`,
			bundle: true,
			target: "firefox115",
			metafile: true
		});
		await this.esbuildContext.rebuild();
	}
	async createTestHtml(tests = []) {
		const setupCode = generateMochaSetup({
			timeout: this.ctx.test.mocha.timeout,
			port: this.port,
			abortOnFail: this.ctx.test.abortOnFail,
			exitOnFinish: !this.ctx.test.watch
		});
		let testFiles = tests;
		if (testFiles.length === 0) testFiles = (await glob(`**/*.{spec,test}.js`, { cwd: `${TESTER_PLUGIN_DIR}/content` })).sort();
		const html = generateHtml(setupCode, testFiles);
		await outputFile(`${TESTER_PLUGIN_DIR}/content/index.xhtml`, html);
	}
};
/**
* Determines which test files are impacted by a given changed file based on the esbuild metafile.
*
* This function analyzes the build metadata to find test files that depend on the changed file
* either directly as an entry point or indirectly as an input. It is useful in a watch mode setup
* to selectively rerun only the affected tests.
*
* @param {string} changedFilePath - The file path of the changed source file.
* @param {BuildResult["metafile"]} buildMetadata - The esbuild metafile containing dependency information.
* @returns {string[]} An array of impacted test file paths that need to be re-executed.
*/
function findImpactedTests(changedFilePath, buildMetadata) {
	if (!buildMetadata) return [];
	const resolvedChangedFile = resolve(changedFilePath);
	const impactedTestFiles = /* @__PURE__ */ new Set();
	for (const [outputFilePath, outputInfo] of Object.entries(buildMetadata.outputs)) {
		const testFilePath = outputFilePath.replace(`${TESTER_PLUGIN_DIR}/content/`, "");
		if (Object.keys(outputInfo.inputs).some((inputPath) => resolve(inputPath) === resolvedChangedFile)) impactedTestFiles.add(testFilePath);
	}
	return [...impactedTestFiles];
}
//#endregion
//#region src/core/tester/index.ts
var Test = class extends Base {
	builder;
	zotero;
	reporter = new TestHttpReporter();
	testBundler;
	constructor(ctx) {
		super(ctx);
		process$1.env.NODE_ENV ??= "test";
		this.builder = new Build(ctx);
		if (isCI) {
			this.ctx.test.headless = true;
			this.ctx.test.watch = false;
		}
	}
	async run() {
		await emptyDir(TESTER_PROFILE_DIR);
		await emptyDir(TESTER_DATA_DIR);
		await emptyDir(TESTER_PLUGIN_DIR);
		await this.ctx.hooks.callHook("test:init", this.ctx);
		await this.builder.run();
		await this.ctx.hooks.callHook("test:prebuild", this.ctx);
		await this.reporter.start();
		this.testBundler = new TestBundler(this.ctx, this.reporter.port);
		await this.testBundler.generate();
		await this.ctx.hooks.callHook("test:bundleTests", this.ctx);
		await this.startZotero();
		await this.ctx.hooks.callHook("test:run", this.ctx);
		if (this.ctx.test.watch) this.watch();
	}
	async watch() {
		const source = toArray(this.ctx.source).map((p) => resolve(p));
		const tests = toArray(this.ctx.test.entries).map((p) => resolve(p));
		function isSource(_path) {
			const path = resolve(_path);
			const isSource = source.find((s) => path.match(s)) || false;
			tests.find((t) => path.match(t));
			return isSource;
		}
		watch([this.ctx.source, this.ctx.test.entries].flat(), this.ctx.watchIgnore, {
			onChange: async (path) => {
				if (isSource(path)) {
					await this.builder.run();
					await this.testBundler?.regenerate(path);
					await this.zotero?.reloadAllPlugins();
				} else {
					await this.testBundler?.regenerate(path);
					await this.zotero?.reloadTemporaryPluginBySourceDir(TESTER_PLUGIN_DIR);
				}
			},
			onAdd: async (path) => {
				if (isSource(path)) {
					await this.builder.run();
					await this.testBundler?.regenerate(path);
					await this.zotero?.reloadAllPlugins();
				} else {
					await this.testBundler?.generate();
					await this.zotero?.reloadTemporaryPluginBySourceDir(TESTER_PLUGIN_DIR);
				}
			}
		});
	}
	async startZotero() {
		if (this.ctx.test.headless) await prepareHeadless();
		this.zotero = new ZoteroRunner({
			binary: {
				path: this.zoteroBinPath,
				devtools: this.ctx.server.devtools,
				args: this.ctx.server.startArgs
			},
			profile: {
				path: TESTER_PROFILE_DIR,
				dataDir: TESTER_DATA_DIR,
				customPrefs: this.prefs
			},
			plugins: { list: [{
				id: this.ctx.id,
				sourceDir: join(this.ctx.dist, "addon")
			}, {
				id: TESTER_PLUGIN_ID,
				sourceDir: TESTER_PLUGIN_DIR
			}] }
		});
		await this.zotero.run();
		this.zotero.zotero?.on("close", () => this.onZoteroExit());
	}
	onZoteroExit = () => {
		this.reporter.stop();
		this.ctx.hooks.callHook("test:exit", this.ctx);
		if (this.reporter.failed) process$1.exit(1);
		else process$1.exit(0);
	};
	exit = (code) => {
		if (code === "SIGINT") this.logger.info("Tester shutdown by user request");
		this.reporter.stop();
		this.zotero?.exit();
		this.ctx.hooks.callHook("test:exit", this.ctx);
		process$1.exit();
	};
	get zoteroBinPath() {
		if (!process$1.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH) throw new Error("No Zotero Found.");
		return process$1.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH;
	}
	get prefs() {
		return Object.assign({
			"extensions.experiments.enabled": true,
			"extensions.autoDisableScopes": 0,
			"devtools.debugger.remote-enabled": true,
			"devtools.debugger.remote-websocket": true,
			"devtools.debugger.prompt-connection": false,
			"app.update.enabled": false,
			"extensions.zotero.sync.server.compressData": false,
			"extensions.zotero.automaticScraperUpdates": false,
			"extensions.zotero.debug.log": 5,
			"extensions.zotero.debug.level": 5,
			"extensions.zotero.debug.time": 5,
			"extensions.zotero.firstRun.skipFirefoxProfileAccessCheck": true,
			"extensions.zotero.firstRunGuidance": false,
			"extensions.zotero.firstRun2": false,
			"extensions.zotero.reportTranslationFailure": false,
			"extensions.zotero.httpServer.enabled": true,
			"extensions.zotero.httpServer.port": 23124,
			"extensions.zotero.httpServer.localAPI.enabled": true,
			"extensions.zotero.backup.numBackups": 0,
			"extensions.zotero.sync.autoSync": false,
			"extensions.zoteroMacWordIntegration.installed": true,
			"extensions.zoteroMacWordIntegration.skipInstallation": true,
			"extensions.zoteroWinWordIntegration.skipInstallation": true,
			"extensions.zoteroOpenOfficeIntegration.skipInstallation": true
		}, this.ctx.test.prefs || {});
	}
};
//#endregion
//#region src/index.ts
const Config = {
	defineConfig,
	loadConfig: loadConfig$1
};
//#endregion
export { Release as a, ExitSignals as i, Test as n, Build as o, Serve as r, defineConfig as s, Config as t };
