// zotero-plugin-scaffold/dist/shared/scaffold-replace-BTU0g6CT.mjs
import { isPlainObject } from "es-toolkit";
import process from "node:process";
import readline from "node:readline";
import { styleText } from "node:util";
import { isDebug } from "std-env";
import { glob } from "tinyglobby";
import { readFile, writeFile } from "node:fs/promises";
//#region src/utils/logger.ts
/**
* Log level enumeration
*/
let LOG_LEVEL = /* @__PURE__ */ function(LOG_LEVEL) {
	LOG_LEVEL[LOG_LEVEL["TRACE"] = 0] = "TRACE";
	LOG_LEVEL[LOG_LEVEL["DEBUG"] = 1] = "DEBUG";
	LOG_LEVEL[LOG_LEVEL["INFO"] = 2] = "INFO";
	LOG_LEVEL[LOG_LEVEL["WARN"] = 3] = "WARN";
	LOG_LEVEL[LOG_LEVEL["ERROR"] = 4] = "ERROR";
	return LOG_LEVEL;
}({});
const SYMBOLS = {
	SUCCESS: styleText("green", "✔"),
	INFO: styleText("blue", "ℹ"),
	FAIL: styleText("red", "✖"),
	TIP: styleText("blue", "→"),
	ERROR: styleText("bgRed", " ERROR "),
	WARN: styleText("bgYellow", " WARN "),
	DEBUG: styleText("gray", "⚙"),
	NONE: ""
};
const DEFAULT_OPTIONS = {
	SPACE: 0,
	LEVEL: LOG_LEVEL.INFO,
	SYMBOL: "",
	NEW_LINE: false
};
/**
* Centralized log method configuration
*/
const LOG_METHODS_CONFIG = {
	error: {
		level: LOG_LEVEL.ERROR,
		symbol: SYMBOLS.ERROR,
		wrapNewLine: true
	},
	warn: {
		level: LOG_LEVEL.WARN,
		symbol: SYMBOLS.WARN,
		wrapNewLine: true
	},
	tip: {
		level: LOG_LEVEL.INFO,
		symbol: SYMBOLS.TIP
	},
	info: {
		level: LOG_LEVEL.INFO,
		symbol: SYMBOLS.INFO
	},
	debug: {
		level: LOG_LEVEL.DEBUG,
		symbol: SYMBOLS.DEBUG
	},
	success: {
		level: LOG_LEVEL.INFO,
		symbol: SYMBOLS.SUCCESS
	},
	ready: {
		level: LOG_LEVEL.INFO,
		symbol: SYMBOLS.SUCCESS,
		wrapNewLine: true
	},
	fail: {
		level: LOG_LEVEL.ERROR,
		symbol: SYMBOLS.FAIL
	},
	log: {
		level: LOG_LEVEL.INFO,
		symbol: SYMBOLS.NONE
	}
};
const logger = class Logger {
	static instance;
	currentLogLevel;
	constructor(level) {
		this.currentLogLevel = this.determineLogLevel(level);
	}
	/**
	* Determine the appropriate log level
	*/
	determineLogLevel(level) {
		if (isDebug) return LOG_LEVEL.TRACE;
		const envLevel = process.env.ZOTERO_PLUGIN_LOG_LEVEL;
		return envLevel ? LOG_LEVEL[envLevel] : level ?? LOG_LEVEL.INFO;
	}
	static getInstance() {
		if (!Logger.instance) Logger.instance = new Logger();
		return Logger.instance;
	}
	setLogLevel(level) {
		this.currentLogLevel = LOG_LEVEL[level];
	}
	get level() {
		return this.currentLogLevel;
	}
	/**
	* Generic log formatting logic
	*/
	formatContent(content) {
		if (typeof content === "string") return content;
		if (content instanceof Error) return this.formatError(content);
		if (isPlainObject(content)) return JSON.stringify(content, null, 2);
		return String(content);
	}
	formatError(error) {
		return `${styleText("red", error.name)}: ${styleText("red", error.message)}\n${error.stack}`;
	}
	/**
	* Core logging method
	*/
	logInternal(content, config, options = {}) {
		if (this.currentLogLevel > config.level) return;
		const { space = DEFAULT_OPTIONS.SPACE, newLine = DEFAULT_OPTIONS.NEW_LINE } = options;
		const formattedContent = this.formatContent(content);
		const output = [
			"  ".repeat(space),
			config.symbol,
			formattedContent
		].join(" ");
		if (config.wrapNewLine) this.newLine();
		console.log(output);
		if (newLine || config.wrapNewLine) this.newLine();
	}
	error(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.error, options);
	}
	warn(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.warn, options);
	}
	tip(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.tip, options);
	}
	info(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.info, options);
	}
	debug(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.debug, options);
	}
	success(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.success, options);
	}
	fail(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.fail, options);
	}
	ready(content) {
		this.logInternal(styleText("green", content), LOG_METHODS_CONFIG.success);
	}
	log(content, options) {
		this.logInternal(content, LOG_METHODS_CONFIG.log, options);
	}
	clear() {
		const blank = process.stdout.rows > 2 ? "\n".repeat(process.stdout.rows - 2) : "";
		console.log(blank);
		readline.cursorTo(process.stdout, 0, 0);
		readline.clearScreenDown(process.stdout);
	}
	newLine() {
		console.log();
	}
}.getInstance();
//#endregion
//#region src/utils/string.ts
function dateFormat(fmt, date) {
	let ret;
	const opt = {
		"Y+": date.getFullYear().toString(),
		"m+": (date.getMonth() + 1).toString(),
		"d+": date.getDate().toString(),
		"H+": date.getHours().toString(),
		"M+": date.getMinutes().toString(),
		"S+": date.getSeconds().toString()
	};
	for (const k in opt) {
		ret = new RegExp(`(${k})`).exec(fmt);
		if (ret) fmt = fmt.replace(ret[1], ret[1].length === 1 ? opt[k] : opt[k].padStart(ret[1].length, "0"));
	}
	return fmt;
}
function toArray(value) {
	return Array.isArray(value) ? value : [value];
}
/**
* template is used to replace data by name in template strings.
* The default expression looks for {{name}} to identify names.
*
* Ex. template('Hello, {{name}}', { name: 'ray' })
* Ex. template('Hello, <name>', { name: 'ray' }, /<(.+?)>/g)
*
* @see https://github.com/sodiray/radash/blob/069b26cdd7d62e6ac16a0ad3baa1c9abcca420bc/src/string.ts#L111-L126
*/
function template(str, data, regex = /\{\{(.+?)\}\}/g) {
	return [...str.matchAll(regex)].reduce((acc, match) => {
		return acc.replace(match[0], data[match[1]]);
	}, str);
}
function parseRepoUrl(url) {
	if (!url) throw new Error("Parse repository URL failed.");
	const match = url.match(/:\/\/.+com\/([^/]+)\/([^.]+)\.git$/);
	if (!match) throw new Error("Parse repository URL failed.");
	const [, owner, repo] = match;
	return {
		owner,
		repo
	};
}
//#endregion
//#region src/core/builder/replace.ts
function replace(contents, from, to) {
	const froms = Array.isArray(from) ? from : [from];
	const tos = Array.isArray(to) ? to : Array.from({ length: froms.length }, () => to);
	if (froms.length !== tos.length) throw new Error("The lengths of 'from' and 'to' must be equal");
	return froms.reduce((result, pattern, index) => result.replace(pattern, tos[index]), contents);
}
async function replaceInFile({ files, from, to, isGlob = true }) {
	const paths = isGlob ? await glob(files) : toArray(files);
	await Promise.all(paths.map(async (path) => {
		const contents = await readFile(path, "utf-8");
		const newContents = replace(contents, from, to);
		if (contents !== newContents) await writeFile(path, newContents);
	}));
}
async function replaceDefine(dist, define) {
	const replaceMap = new Map(Object.keys(define).map((key) => [new RegExp(`__${key}__`, "g"), define[key]]));
	logger.debug(`replace map: ${replaceMap}`);
	await replaceInFile({
		files: `${dist}/addon/**/*`,
		from: [...replaceMap.keys()],
		to: [...replaceMap.values()]
	});
}
//#endregion
export { parseRepoUrl as a, LOG_LEVEL as c, dateFormat as i, logger as l, replaceDefine as n, template as o, replaceInFile as r, toArray as s, replace as t };
